{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c3bb10ad-4b8e-4c25-9b1a-6565e45b3499",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: cas-tools in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (1.1.2)\n",
      "Requirement already satisfied: deepmerge==1.1.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from cas-tools) (1.1.0)\n",
      "Requirement already satisfied: marshmallow==3.21.1 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from cas-tools) (3.21.1)\n",
      "Requirement already satisfied: cell-annotation-schema==1.0.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from cas-tools) (1.0.0)\n",
      "Requirement already satisfied: defusedxml==0.7.1 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from cas-tools) (0.7.1)\n",
      "Requirement already satisfied: rdflib==7.0.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from cas-tools) (7.0.0)\n",
      "Requirement already satisfied: openpyxl==3.1.2 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from cas-tools) (3.1.2)\n",
      "Requirement already satisfied: pandas==2.2.1 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from cas-tools) (2.2.1)\n",
      "Requirement already satisfied: dataclasses-json==0.6.4 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from cas-tools) (0.6.4)\n",
      "Requirement already satisfied: numpy==1.26.4 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from cas-tools) (1.26.4)\n",
      "Requirement already satisfied: linkml==1.7.8 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from cas-tools) (1.7.8)\n",
      "Requirement already satisfied: oaklib==0.6.4 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from cas-tools) (0.6.4)\n",
      "Requirement already satisfied: python-dateutil==2.9.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from cas-tools) (2.9.0)\n",
      "Requirement already satisfied: schema-automator==0.5.1 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from cas-tools) (0.5.1)\n",
      "Requirement already satisfied: urllib3<2 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from cas-tools) (1.26.20)\n",
      "Requirement already satisfied: h5py==3.10.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from cas-tools) (3.10.0)\n",
      "Requirement already satisfied: psycopg2-binary==2.9.5 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from cas-tools) (2.9.5)\n",
      "Requirement already satisfied: ordered-set==4.1.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from cas-tools) (4.1.0)\n",
      "Requirement already satisfied: cas-linkml<0.0.15,>=0.0.14 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from cas-tools) (0.0.14)\n",
      "Requirement already satisfied: jsonschema==4.4.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from cas-tools) (4.4.0)\n",
      "Requirement already satisfied: cap-anndata<0.4.0,>=0.3.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from cas-tools) (0.3.1)\n",
      "Requirement already satisfied: ruamel-yaml==0.18.6 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from cas-tools) (0.18.6)\n",
      "Requirement already satisfied: cellxgene-census==1.15.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from cas-tools) (1.15.0)\n",
      "Requirement already satisfied: requests==2.32.3 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from cas-tools) (2.32.3)\n",
      "Requirement already satisfied: anndata==0.10.5 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from cas-tools) (0.10.5)\n",
      "Requirement already satisfied: scipy>1.4 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from anndata==0.10.5->cas-tools) (1.13.1)\n",
      "Requirement already satisfied: exceptiongroup in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from anndata==0.10.5->cas-tools) (1.2.2)\n",
      "Requirement already satisfied: natsort in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from anndata==0.10.5->cas-tools) (8.4.0)\n",
      "Requirement already satisfied: packaging>=20 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from anndata==0.10.5->cas-tools) (24.2)\n",
      "Requirement already satisfied: array-api-compat in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from anndata==0.10.5->cas-tools) (1.11.2)\n",
      "Requirement already satisfied: tiledbsoma~=1.11.4 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from cellxgene-census==1.15.0->cas-tools) (1.11.4)\n",
      "Requirement already satisfied: typing-extensions in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from cellxgene-census==1.15.0->cas-tools) (4.12.2)\n",
      "Requirement already satisfied: s3fs>=2021.06.1 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from cellxgene-census==1.15.0->cas-tools) (2025.5.1)\n",
      "Requirement already satisfied: typing-inspect<1,>=0.4.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from dataclasses-json==0.6.4->cas-tools) (0.9.0)\n",
      "Requirement already satisfied: attrs>=17.4.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from jsonschema==4.4.0->cas-tools) (25.1.0)\n",
      "Requirement already satisfied: pyrsistent!=0.17.0,!=0.17.1,!=0.17.2,>=0.14.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from jsonschema==4.4.0->cas-tools) (0.20.0)\n",
      "Requirement already satisfied: parse in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from linkml==1.7.8->cas-tools) (1.20.2)\n",
      "Requirement already satisfied: jsonasobj2<2.0.0,>=1.0.3 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from linkml==1.7.8->cas-tools) (1.0.4)\n",
      "Requirement already satisfied: pyshex>=0.7.20 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from linkml==1.7.8->cas-tools) (0.8.1)\n",
      "Requirement already satisfied: hbreader in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from linkml==1.7.8->cas-tools) (0.9.1)\n",
      "Requirement already satisfied: watchdog>=0.9.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from linkml==1.7.8->cas-tools) (6.0.0)\n",
      "Requirement already satisfied: linkml-dataops in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from linkml==1.7.8->cas-tools) (0.1.0)\n",
      "Requirement already satisfied: graphviz>=0.10.1 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from linkml==1.7.8->cas-tools) (0.20.3)\n",
      "Requirement already satisfied: antlr4-python3-runtime<4.10,>=4.9.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from linkml==1.7.8->cas-tools) (4.9.3)\n",
      "Requirement already satisfied: pydantic<3.0.0,>=1.0.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from linkml==1.7.8->cas-tools) (2.11.5)\n",
      "Requirement already satisfied: isodate>=0.6.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from linkml==1.7.8->cas-tools) (0.6.1)\n",
      "Requirement already satisfied: pyshexc>=0.8.3 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from linkml==1.7.8->cas-tools) (0.9.1)\n",
      "Requirement already satisfied: prefixmaps>=0.2.2 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from linkml==1.7.8->cas-tools) (0.2.6)\n",
      "Requirement already satisfied: sqlalchemy>=1.4.31 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from linkml==1.7.8->cas-tools) (2.0.41)\n",
      "Requirement already satisfied: pyjsg>=0.11.6 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from linkml==1.7.8->cas-tools) (0.11.10)\n",
      "Requirement already satisfied: pyyaml in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from linkml==1.7.8->cas-tools) (6.0.2)\n",
      "Requirement already satisfied: click>=7.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from linkml==1.7.8->cas-tools) (8.1.8)\n",
      "Requirement already satisfied: linkml-runtime>=1.7.4 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from linkml==1.7.8->cas-tools) (1.8.3)\n",
      "Requirement already satisfied: prefixcommons>=0.1.7 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from linkml==1.7.8->cas-tools) (0.1.12)\n",
      "Requirement already satisfied: jinja2>=3.1.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from linkml==1.7.8->cas-tools) (3.1.5)\n",
      "Requirement already satisfied: pystow>=0.5.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from oaklib==0.6.4->cas-tools) (0.7.0)\n",
      "Requirement already satisfied: curies>=0.6.6 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from oaklib==0.6.4->cas-tools) (0.10.19)\n",
      "Requirement already satisfied: pronto>=2.5.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from oaklib==0.6.4->cas-tools) (2.7.0)\n",
      "Requirement already satisfied: airium>=0.2.5 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from oaklib==0.6.4->cas-tools) (0.2.6)\n",
      "Requirement already satisfied: networkx>=2.7.1 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from oaklib==0.6.4->cas-tools) (3.2.1)\n",
      "Requirement already satisfied: class-resolver>=0.4.2 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from oaklib==0.6.4->cas-tools) (0.6.0)\n",
      "Requirement already satisfied: linkml-renderer>=0.3.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from oaklib==0.6.4->cas-tools) (0.3.1)\n",
      "Requirement already satisfied: requests-cache<2.0.0,>=1.0.1 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from oaklib==0.6.4->cas-tools) (1.2.1)\n",
      "Requirement already satisfied: jsonlines<5.0.0,>=4.0.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from oaklib==0.6.4->cas-tools) (4.0.0)\n",
      "Requirement already satisfied: funowl>=0.2.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from oaklib==0.6.4->cas-tools) (0.2.3)\n",
      "Requirement already satisfied: eutils>=0.6.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from oaklib==0.6.4->cas-tools) (0.6.0)\n",
      "Requirement already satisfied: pysolr<4.0.0,>=3.9.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from oaklib==0.6.4->cas-tools) (3.10.0)\n",
      "Requirement already satisfied: ratelimit>=2.2.1 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from oaklib==0.6.4->cas-tools) (2.2.1)\n",
      "Requirement already satisfied: semsql>=0.3.1 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from oaklib==0.6.4->cas-tools) (0.4.0)\n",
      "Requirement already satisfied: kgcl-rdflib==0.5.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from oaklib==0.6.4->cas-tools) (0.5.0)\n",
      "Requirement already satisfied: kgcl-schema<0.7.0,>=0.6.8 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from oaklib==0.6.4->cas-tools) (0.6.9)\n",
      "Requirement already satisfied: ontoportal-client>=0.0.3 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from oaklib==0.6.4->cas-tools) (0.0.4)\n",
      "Requirement already satisfied: appdirs>=1.4.4 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from oaklib==0.6.4->cas-tools) (1.4.4)\n",
      "Requirement already satisfied: ndex2<4.0.0,>=3.5.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from oaklib==0.6.4->cas-tools) (3.10.0)\n",
      "Requirement already satisfied: sssom<0.5.0,>=0.4.4 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from oaklib==0.6.4->cas-tools) (0.4.11)\n",
      "Requirement already satisfied: tenacity<9.0.0,>=8.2.3 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from oaklib==0.6.4->cas-tools) (8.5.0)\n",
      "Requirement already satisfied: SPARQLWrapper in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from oaklib==0.6.4->cas-tools) (2.0.0)\n",
      "Requirement already satisfied: ols-client>=0.1.1 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from oaklib==0.6.4->cas-tools) (0.1.4)\n",
      "Requirement already satisfied: et-xmlfile in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from openpyxl==3.1.2->cas-tools) (2.0.0)\n",
      "Requirement already satisfied: pytz>=2020.1 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from pandas==2.2.1->cas-tools) (2025.1)\n",
      "Requirement already satisfied: tzdata>=2022.7 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from pandas==2.2.1->cas-tools) (2025.1)\n",
      "Requirement already satisfied: six>=1.5 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from python-dateutil==2.9.0->cas-tools) (1.17.0)\n",
      "Requirement already satisfied: pyparsing<4,>=2.1.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from rdflib==7.0.0->cas-tools) (3.2.1)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from requests==2.32.3->cas-tools) (3.4.1)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from requests==2.32.3->cas-tools) (3.10)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from requests==2.32.3->cas-tools) (2025.1.31)\n",
      "Requirement already satisfied: ruamel.yaml.clib>=0.2.7 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from ruamel-yaml==0.18.6->cas-tools) (0.2.12)\n",
      "Requirement already satisfied: tomlkit>=0.11.4 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from schema-automator==0.5.1->cas-tools) (0.13.2)\n",
      "Requirement already satisfied: schemasheets>=0.1.24 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from schema-automator==0.5.1->cas-tools) (0.4.0)\n",
      "Requirement already satisfied: xmltodict<0.14.0,>=0.13.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from schema-automator==0.5.1->cas-tools) (0.13.0)\n",
      "Requirement already satisfied: duckdb<0.11.0,>=0.10.1 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from schema-automator==0.5.1->cas-tools) (0.10.3)\n",
      "Requirement already satisfied: strsimpy>=0.2.1 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from schema-automator==0.5.1->cas-tools) (0.2.1)\n",
      "Requirement already satisfied: mkdocs>=1.2.3 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from schema-automator==0.5.1->cas-tools) (1.6.1)\n",
      "Requirement already satisfied: jsonpatch>=1.32 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from schema-automator==0.5.1->cas-tools) (1.33)\n",
      "Requirement already satisfied: quantulum3>=0.7.9 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from schema-automator==0.5.1->cas-tools) (0.9.2)\n",
      "Requirement already satisfied: click-log>=0.4.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from schema-automator==0.5.1->cas-tools) (0.4.0)\n",
      "Requirement already satisfied: inflect>=6.0.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from schema-automator==0.5.1->cas-tools) (7.5.0)\n",
      "Requirement already satisfied: click-default-group<2.0.0,>=1.2.4 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from schema-automator==0.5.1->cas-tools) (1.2.4)\n",
      "Requirement already satisfied: pandera>=0.12.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from schema-automator==0.5.1->cas-tools) (0.24.0)\n",
      "Requirement already satisfied: dacite<2.0.0,>=1.8.1 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from cas-linkml<0.0.15,>=0.0.14->cas-tools) (1.9.2)\n",
      "Requirement already satisfied: importlib-metadata>3.6 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from class-resolver>=0.4.2->oaklib==0.6.4->cas-tools) (8.6.1)\n",
      "Requirement already satisfied: pytrie in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from curies>=0.6.6->oaklib==0.6.4->cas-tools) (0.4.0)\n",
      "Requirement already satisfied: eval-type-backport in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from curies>=0.6.6->oaklib==0.6.4->cas-tools) (0.2.2)\n",
      "Requirement already satisfied: lxml in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from eutils>=0.6.0->oaklib==0.6.4->cas-tools) (5.4.0)\n",
      "Requirement already satisfied: rdflib-shim in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from funowl>=0.2.0->oaklib==0.6.4->cas-tools) (1.0.3)\n",
      "Requirement already satisfied: bcp47 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from funowl>=0.2.0->oaklib==0.6.4->cas-tools) (0.1.0)\n",
      "Requirement already satisfied: jsonasobj in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from funowl>=0.2.0->oaklib==0.6.4->cas-tools) (1.3.1)\n",
      "Requirement already satisfied: rfc3987 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from funowl>=0.2.0->oaklib==0.6.4->cas-tools) (1.3.8)\n",
      "Requirement already satisfied: zipp>=3.20 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from importlib-metadata>3.6->class-resolver>=0.4.2->oaklib==0.6.4->cas-tools) (3.21.0)\n",
      "Requirement already satisfied: more_itertools>=8.5.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from inflect>=6.0.0->schema-automator==0.5.1->cas-tools) (10.7.0)\n",
      "Requirement already satisfied: typeguard>=4.0.1 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from inflect>=6.0.0->schema-automator==0.5.1->cas-tools) (4.4.2)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from jinja2>=3.1.0->linkml==1.7.8->cas-tools) (3.0.2)\n",
      "Requirement already satisfied: jsonpointer>=1.9 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from jsonpatch>=1.32->schema-automator==0.5.1->cas-tools) (3.0.0)\n",
      "Requirement already satisfied: rfc3339-validator in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from jsonschema==4.4.0->cas-tools) (0.1.4)\n",
      "Requirement already satisfied: fqdn in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from jsonschema==4.4.0->cas-tools) (1.5.1)\n",
      "Requirement already satisfied: uri-template in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from jsonschema==4.4.0->cas-tools) (1.3.0)\n",
      "Requirement already satisfied: webcolors>=1.11 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from jsonschema==4.4.0->cas-tools) (24.11.1)\n",
      "Requirement already satisfied: isoduration in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from jsonschema==4.4.0->cas-tools) (20.11.0)\n",
      "Requirement already satisfied: lark>=1.1.2 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from kgcl-schema<0.7.0,>=0.6.8->oaklib==0.6.4->cas-tools) (1.2.2)\n",
      "Requirement already satisfied: deprecated in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from linkml-runtime>=1.7.4->linkml==1.7.8->cas-tools) (1.2.18)\n",
      "Requirement already satisfied: json-flattener>=0.1.9 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from linkml-runtime>=1.7.4->linkml==1.7.8->cas-tools) (0.1.9)\n",
      "Requirement already satisfied: markdown>=3.3.6 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from mkdocs>=1.2.3->schema-automator==0.5.1->cas-tools) (3.8)\n",
      "Requirement already satisfied: pathspec>=0.11.1 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from mkdocs>=1.2.3->schema-automator==0.5.1->cas-tools) (0.12.1)\n",
      "Requirement already satisfied: pyyaml-env-tag>=0.1 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from mkdocs>=1.2.3->schema-automator==0.5.1->cas-tools) (1.1)\n",
      "Requirement already satisfied: mergedeep>=1.3.4 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from mkdocs>=1.2.3->schema-automator==0.5.1->cas-tools) (1.3.4)\n",
      "Requirement already satisfied: ghp-import>=1.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from mkdocs>=1.2.3->schema-automator==0.5.1->cas-tools) (2.1.0)\n",
      "Requirement already satisfied: mkdocs-get-deps>=0.2.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from mkdocs>=1.2.3->schema-automator==0.5.1->cas-tools) (0.2.0)\n",
      "Requirement already satisfied: platformdirs>=2.2.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from mkdocs-get-deps>=0.2.0->mkdocs>=1.2.3->schema-automator==0.5.1->cas-tools) (4.3.6)\n",
      "Requirement already satisfied: requests-toolbelt in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from ndex2<4.0.0,>=3.5.0->oaklib==0.6.4->cas-tools) (1.0.0)\n",
      "Requirement already satisfied: ijson in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from ndex2<4.0.0,>=3.5.0->oaklib==0.6.4->cas-tools) (3.4.0)\n",
      "Requirement already satisfied: more-click in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from ols-client>=0.1.1->oaklib==0.6.4->cas-tools) (0.1.2)\n",
      "Requirement already satisfied: pytest-logging<2016.0.0,>=2015.11.4 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from prefixcommons>=0.1.7->linkml==1.7.8->cas-tools) (2015.11.4)\n",
      "Requirement already satisfied: fastobo~=0.13.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from pronto>=2.5.0->oaklib==0.6.4->cas-tools) (0.13.0)\n",
      "Requirement already satisfied: chardet~=5.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from pronto>=2.5.0->oaklib==0.6.4->cas-tools) (5.2.0)\n",
      "Requirement already satisfied: annotated-types>=0.6.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from pydantic<3.0.0,>=1.0.0->linkml==1.7.8->cas-tools) (0.7.0)\n",
      "Requirement already satisfied: typing-inspection>=0.4.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from pydantic<3.0.0,>=1.0.0->linkml==1.7.8->cas-tools) (0.4.1)\n",
      "Requirement already satisfied: pydantic-core==2.33.2 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from pydantic<3.0.0,>=1.0.0->linkml==1.7.8->cas-tools) (2.33.2)\n",
      "Requirement already satisfied: shexjsg>=0.8.2 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from pyshex>=0.7.20->linkml==1.7.8->cas-tools) (0.8.2)\n",
      "Requirement already satisfied: cfgraph>=0.2.1 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from pyshex>=0.7.20->linkml==1.7.8->cas-tools) (0.2.1)\n",
      "Requirement already satisfied: sparqlslurper>=0.5.1 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from pyshex>=0.7.20->linkml==1.7.8->cas-tools) (0.5.1)\n",
      "Requirement already satisfied: setuptools in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from pysolr<4.0.0,>=3.9.0->oaklib==0.6.4->cas-tools) (58.0.4)\n",
      "Requirement already satisfied: tqdm in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from pystow>=0.5.0->oaklib==0.6.4->cas-tools) (4.67.1)\n",
      "Requirement already satisfied: pytest>=2.8.1 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from pytest-logging<2016.0.0,>=2015.11.4->prefixcommons>=0.1.7->linkml==1.7.8->cas-tools) (8.3.5)\n",
      "Requirement already satisfied: tomli>=1 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from pytest>=2.8.1->pytest-logging<2016.0.0,>=2015.11.4->prefixcommons>=0.1.7->linkml==1.7.8->cas-tools) (2.2.1)\n",
      "Requirement already satisfied: pluggy<2,>=1.5 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from pytest>=2.8.1->pytest-logging<2016.0.0,>=2015.11.4->prefixcommons>=0.1.7->linkml==1.7.8->cas-tools) (1.6.0)\n",
      "Requirement already satisfied: iniconfig in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from pytest>=2.8.1->pytest-logging<2016.0.0,>=2015.11.4->prefixcommons>=0.1.7->linkml==1.7.8->cas-tools) (2.1.0)\n",
      "Requirement already satisfied: num2words in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from quantulum3>=0.7.9->schema-automator==0.5.1->cas-tools) (0.5.14)\n",
      "Requirement already satisfied: url-normalize>=1.4 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from requests-cache<2.0.0,>=1.0.1->oaklib==0.6.4->cas-tools) (2.2.1)\n",
      "Requirement already satisfied: cattrs>=22.2 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from requests-cache<2.0.0,>=1.0.1->oaklib==0.6.4->cas-tools) (24.1.3)\n",
      "Requirement already satisfied: fsspec==2025.5.1 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from s3fs>=2021.06.1->cellxgene-census==1.15.0->cas-tools) (2025.5.1)\n",
      "Requirement already satisfied: aiobotocore<3.0.0,>=2.5.4 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from s3fs>=2021.06.1->cellxgene-census==1.15.0->cas-tools) (2.22.0)\n",
      "Requirement already satisfied: aiohttp!=4.0.0a0,!=4.0.0a1 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from s3fs>=2021.06.1->cellxgene-census==1.15.0->cas-tools) (3.12.4)\n",
      "Requirement already satisfied: jmespath<2.0.0,>=0.7.1 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from aiobotocore<3.0.0,>=2.5.4->s3fs>=2021.06.1->cellxgene-census==1.15.0->cas-tools) (1.0.1)\n",
      "Requirement already satisfied: multidict<7.0.0,>=6.0.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from aiobotocore<3.0.0,>=2.5.4->s3fs>=2021.06.1->cellxgene-census==1.15.0->cas-tools) (6.4.4)\n",
      "Requirement already satisfied: wrapt<2.0.0,>=1.10.10 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from aiobotocore<3.0.0,>=2.5.4->s3fs>=2021.06.1->cellxgene-census==1.15.0->cas-tools) (1.17.2)\n",
      "Requirement already satisfied: botocore<1.37.4,>=1.37.2 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from aiobotocore<3.0.0,>=2.5.4->s3fs>=2021.06.1->cellxgene-census==1.15.0->cas-tools) (1.37.3)\n",
      "Requirement already satisfied: aioitertools<1.0.0,>=0.5.1 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from aiobotocore<3.0.0,>=2.5.4->s3fs>=2021.06.1->cellxgene-census==1.15.0->cas-tools) (0.12.0)\n",
      "Requirement already satisfied: aiosignal>=1.1.2 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs>=2021.06.1->cellxgene-census==1.15.0->cas-tools) (1.3.2)\n",
      "Requirement already satisfied: aiohappyeyeballs>=2.5.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs>=2021.06.1->cellxgene-census==1.15.0->cas-tools) (2.6.1)\n",
      "Requirement already satisfied: yarl<2.0,>=1.17.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs>=2021.06.1->cellxgene-census==1.15.0->cas-tools) (1.20.0)\n",
      "Requirement already satisfied: async-timeout<6.0,>=4.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs>=2021.06.1->cellxgene-census==1.15.0->cas-tools) (5.0.1)\n",
      "Requirement already satisfied: propcache>=0.2.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs>=2021.06.1->cellxgene-census==1.15.0->cas-tools) (0.3.1)\n",
      "Requirement already satisfied: frozenlist>=1.1.1 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs>=2021.06.1->cellxgene-census==1.15.0->cas-tools) (1.6.0)\n",
      "Requirement already satisfied: ontodev-cogs<0.4.0,>=0.3.3 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from schemasheets>=0.1.24->schema-automator==0.5.1->cas-tools) (0.3.3)\n",
      "Requirement already satisfied: bioregistry>0.5.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from schemasheets>=0.1.24->schema-automator==0.5.1->cas-tools) (0.12.14)\n",
      "Requirement already satisfied: gspread-formatting in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from ontodev-cogs<0.4.0,>=0.3.3->schemasheets>=0.1.24->schema-automator==0.5.1->cas-tools) (1.2.1)\n",
      "Requirement already satisfied: google in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from ontodev-cogs<0.4.0,>=0.3.3->schemasheets>=0.1.24->schema-automator==0.5.1->cas-tools) (3.0.0)\n",
      "Requirement already satisfied: google-api-python-client in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from ontodev-cogs<0.4.0,>=0.3.3->schemasheets>=0.1.24->schema-automator==0.5.1->cas-tools) (2.170.0)\n",
      "Requirement already satisfied: gspread in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from ontodev-cogs<0.4.0,>=0.3.3->schemasheets>=0.1.24->schema-automator==0.5.1->cas-tools) (6.2.1)\n",
      "Requirement already satisfied: tabulate in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from ontodev-cogs<0.4.0,>=0.3.3->schemasheets>=0.1.24->schema-automator==0.5.1->cas-tools) (0.9.0)\n",
      "Requirement already satisfied: termcolor in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from ontodev-cogs<0.4.0,>=0.3.3->schemasheets>=0.1.24->schema-automator==0.5.1->cas-tools) (3.1.0)\n",
      "Requirement already satisfied: daff in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from ontodev-cogs<0.4.0,>=0.3.3->schemasheets>=0.1.24->schema-automator==0.5.1->cas-tools) (1.4.2)\n",
      "Requirement already satisfied: email-validator>=2.0.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from pydantic<3.0.0,>=1.0.0->linkml==1.7.8->cas-tools) (2.2.0)\n",
      "Requirement already satisfied: dnspython>=2.0.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from email-validator>=2.0.0->pydantic<3.0.0,>=1.0.0->linkml==1.7.8->cas-tools) (2.7.0)\n",
      "Requirement already satisfied: SQLAlchemy-Utils<0.39.0,>=0.38.2 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from semsql>=0.3.1->oaklib==0.6.4->cas-tools) (0.38.3)\n",
      "Requirement already satisfied: pansql[pansql]>=0.0.1 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from sssom<0.5.0,>=0.4.4->oaklib==0.6.4->cas-tools) (0.0.1)\n",
      "Requirement already satisfied: importlib-resources<7.0.0,>=6.1.1 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from sssom<0.5.0,>=0.4.4->oaklib==0.6.4->cas-tools) (6.5.2)\n",
      "Requirement already satisfied: sssom-schema<0.16.0,>=0.15.2 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from sssom<0.5.0,>=0.4.4->oaklib==0.6.4->cas-tools) (0.15.2)\n",
      "Requirement already satisfied: deprecation<3.0.0,>=2.1.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from sssom<0.5.0,>=0.4.4->oaklib==0.6.4->cas-tools) (2.1.0)\n",
      "Requirement already satisfied: validators>=0.20.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from sssom<0.5.0,>=0.4.4->oaklib==0.6.4->cas-tools) (0.35.0)\n",
      "\u001b[33mWARNING: networkx 3.2.1 does not provide the extra 'networkx'\u001b[0m\n",
      "\u001b[33mWARNING: pansql 0.0.1 does not provide the extra 'pansql'\u001b[0m\n",
      "Requirement already satisfied: somacore==1.0.11 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from tiledbsoma~=1.11.4->cellxgene-census==1.15.0->cas-tools) (1.0.11)\n",
      "Requirement already satisfied: pyarrow<13.0.0,>=9.0.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from tiledbsoma~=1.11.4->cellxgene-census==1.15.0->cas-tools) (12.0.1)\n",
      "Requirement already satisfied: scanpy>=1.9.2 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from tiledbsoma~=1.11.4->cellxgene-census==1.15.0->cas-tools) (1.10.3)\n",
      "Requirement already satisfied: pyarrow-hotfix in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from tiledbsoma~=1.11.4->cellxgene-census==1.15.0->cas-tools) (0.7)\n",
      "Requirement already satisfied: tiledb~=0.29.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from tiledbsoma~=1.11.4->cellxgene-census==1.15.0->cas-tools) (0.29.1)\n",
      "Requirement already satisfied: numba>=0.58.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from tiledbsoma~=1.11.4->cellxgene-census==1.15.0->cas-tools) (0.60.0)\n",
      "Requirement already satisfied: llvmlite<0.44,>=0.43.0dev0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from numba>=0.58.0->tiledbsoma~=1.11.4->cellxgene-census==1.15.0->cas-tools) (0.43.0)\n",
      "Requirement already satisfied: scikit-learn>=0.24 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from scanpy>=1.9.2->tiledbsoma~=1.11.4->cellxgene-census==1.15.0->cas-tools) (1.6.1)\n",
      "Requirement already satisfied: seaborn>=0.13 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from scanpy>=1.9.2->tiledbsoma~=1.11.4->cellxgene-census==1.15.0->cas-tools) (0.13.2)\n",
      "Requirement already satisfied: session-info in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from scanpy>=1.9.2->tiledbsoma~=1.11.4->cellxgene-census==1.15.0->cas-tools) (1.0.1)\n",
      "Requirement already satisfied: umap-learn!=0.5.0,>=0.5 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from scanpy>=1.9.2->tiledbsoma~=1.11.4->cellxgene-census==1.15.0->cas-tools) (0.5.7)\n",
      "Requirement already satisfied: patsy in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from scanpy>=1.9.2->tiledbsoma~=1.11.4->cellxgene-census==1.15.0->cas-tools) (1.0.1)\n",
      "Requirement already satisfied: pynndescent>=0.5 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from scanpy>=1.9.2->tiledbsoma~=1.11.4->cellxgene-census==1.15.0->cas-tools) (0.5.13)\n",
      "Requirement already satisfied: matplotlib>=3.6 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from scanpy>=1.9.2->tiledbsoma~=1.11.4->cellxgene-census==1.15.0->cas-tools) (3.9.4)\n",
      "Requirement already satisfied: get-annotations in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from scanpy>=1.9.2->tiledbsoma~=1.11.4->cellxgene-census==1.15.0->cas-tools) (0.1.2)\n",
      "Requirement already satisfied: legacy-api-wrap>=1.4 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from scanpy>=1.9.2->tiledbsoma~=1.11.4->cellxgene-census==1.15.0->cas-tools) (1.4.1)\n",
      "Requirement already satisfied: joblib in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from scanpy>=1.9.2->tiledbsoma~=1.11.4->cellxgene-census==1.15.0->cas-tools) (1.5.1)\n",
      "Requirement already satisfied: statsmodels>=0.13 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from scanpy>=1.9.2->tiledbsoma~=1.11.4->cellxgene-census==1.15.0->cas-tools) (0.14.4)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from matplotlib>=3.6->scanpy>=1.9.2->tiledbsoma~=1.11.4->cellxgene-census==1.15.0->cas-tools) (4.58.1)\n",
      "Requirement already satisfied: pillow>=8 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from matplotlib>=3.6->scanpy>=1.9.2->tiledbsoma~=1.11.4->cellxgene-census==1.15.0->cas-tools) (11.2.1)\n",
      "Requirement already satisfied: cycler>=0.10 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from matplotlib>=3.6->scanpy>=1.9.2->tiledbsoma~=1.11.4->cellxgene-census==1.15.0->cas-tools) (0.12.1)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from matplotlib>=3.6->scanpy>=1.9.2->tiledbsoma~=1.11.4->cellxgene-census==1.15.0->cas-tools) (1.3.0)\n",
      "Requirement already satisfied: kiwisolver>=1.3.1 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from matplotlib>=3.6->scanpy>=1.9.2->tiledbsoma~=1.11.4->cellxgene-census==1.15.0->cas-tools) (1.4.7)\n",
      "Requirement already satisfied: threadpoolctl>=3.1.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from scikit-learn>=0.24->scanpy>=1.9.2->tiledbsoma~=1.11.4->cellxgene-census==1.15.0->cas-tools) (3.6.0)\n",
      "Requirement already satisfied: mypy-extensions>=0.3.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from typing-inspect<1,>=0.4.0->dataclasses-json==0.6.4->cas-tools) (1.1.0)\n",
      "Requirement already satisfied: beautifulsoup4 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from google->ontodev-cogs<0.4.0,>=0.3.3->schemasheets>=0.1.24->schema-automator==0.5.1->cas-tools) (4.13.3)\n",
      "Requirement already satisfied: soupsieve>1.2 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from beautifulsoup4->google->ontodev-cogs<0.4.0,>=0.3.3->schemasheets>=0.1.24->schema-automator==0.5.1->cas-tools) (2.6)\n",
      "Requirement already satisfied: google-auth-httplib2<1.0.0,>=0.2.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from google-api-python-client->ontodev-cogs<0.4.0,>=0.3.3->schemasheets>=0.1.24->schema-automator==0.5.1->cas-tools) (0.2.0)\n",
      "Requirement already satisfied: uritemplate<5,>=3.0.1 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from google-api-python-client->ontodev-cogs<0.4.0,>=0.3.3->schemasheets>=0.1.24->schema-automator==0.5.1->cas-tools) (4.1.1)\n",
      "Requirement already satisfied: google-api-core!=2.0.*,!=2.1.*,!=2.2.*,!=2.3.0,<3.0.0,>=1.31.5 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from google-api-python-client->ontodev-cogs<0.4.0,>=0.3.3->schemasheets>=0.1.24->schema-automator==0.5.1->cas-tools) (2.24.2)\n",
      "Requirement already satisfied: google-auth!=2.24.0,!=2.25.0,<3.0.0,>=1.32.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from google-api-python-client->ontodev-cogs<0.4.0,>=0.3.3->schemasheets>=0.1.24->schema-automator==0.5.1->cas-tools) (2.40.2)\n",
      "Requirement already satisfied: httplib2<1.0.0,>=0.19.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from google-api-python-client->ontodev-cogs<0.4.0,>=0.3.3->schemasheets>=0.1.24->schema-automator==0.5.1->cas-tools) (0.22.0)\n",
      "Requirement already satisfied: googleapis-common-protos<2.0.0,>=1.56.2 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from google-api-core!=2.0.*,!=2.1.*,!=2.2.*,!=2.3.0,<3.0.0,>=1.31.5->google-api-python-client->ontodev-cogs<0.4.0,>=0.3.3->schemasheets>=0.1.24->schema-automator==0.5.1->cas-tools) (1.70.0)\n",
      "Requirement already satisfied: protobuf!=3.20.0,!=3.20.1,!=4.21.0,!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5,<7.0.0,>=3.19.5 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from google-api-core!=2.0.*,!=2.1.*,!=2.2.*,!=2.3.0,<3.0.0,>=1.31.5->google-api-python-client->ontodev-cogs<0.4.0,>=0.3.3->schemasheets>=0.1.24->schema-automator==0.5.1->cas-tools) (6.31.0)\n",
      "Requirement already satisfied: proto-plus<2.0.0,>=1.22.3 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from google-api-core!=2.0.*,!=2.1.*,!=2.2.*,!=2.3.0,<3.0.0,>=1.31.5->google-api-python-client->ontodev-cogs<0.4.0,>=0.3.3->schemasheets>=0.1.24->schema-automator==0.5.1->cas-tools) (1.26.1)\n",
      "Requirement already satisfied: rsa<5,>=3.1.4 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from google-auth!=2.24.0,!=2.25.0,<3.0.0,>=1.32.0->google-api-python-client->ontodev-cogs<0.4.0,>=0.3.3->schemasheets>=0.1.24->schema-automator==0.5.1->cas-tools) (4.9.1)\n",
      "Requirement already satisfied: cachetools<6.0,>=2.0.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from google-auth!=2.24.0,!=2.25.0,<3.0.0,>=1.32.0->google-api-python-client->ontodev-cogs<0.4.0,>=0.3.3->schemasheets>=0.1.24->schema-automator==0.5.1->cas-tools) (5.5.2)\n",
      "Requirement already satisfied: pyasn1-modules>=0.2.1 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from google-auth!=2.24.0,!=2.25.0,<3.0.0,>=1.32.0->google-api-python-client->ontodev-cogs<0.4.0,>=0.3.3->schemasheets>=0.1.24->schema-automator==0.5.1->cas-tools) (0.4.2)\n",
      "Requirement already satisfied: pyasn1<0.7.0,>=0.6.1 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from pyasn1-modules>=0.2.1->google-auth!=2.24.0,!=2.25.0,<3.0.0,>=1.32.0->google-api-python-client->ontodev-cogs<0.4.0,>=0.3.3->schemasheets>=0.1.24->schema-automator==0.5.1->cas-tools) (0.6.1)\n",
      "Requirement already satisfied: google-auth-oauthlib>=0.4.1 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from gspread->ontodev-cogs<0.4.0,>=0.3.3->schemasheets>=0.1.24->schema-automator==0.5.1->cas-tools) (1.2.2)\n",
      "Requirement already satisfied: requests-oauthlib>=0.7.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from google-auth-oauthlib>=0.4.1->gspread->ontodev-cogs<0.4.0,>=0.3.3->schemasheets>=0.1.24->schema-automator==0.5.1->cas-tools) (2.0.0)\n",
      "Requirement already satisfied: oauthlib>=3.0.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from requests-oauthlib>=0.7.0->google-auth-oauthlib>=0.4.1->gspread->ontodev-cogs<0.4.0,>=0.3.3->schemasheets>=0.1.24->schema-automator==0.5.1->cas-tools) (3.2.2)\n",
      "Requirement already satisfied: arrow>=0.15.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from isoduration->jsonschema==4.4.0->cas-tools) (1.3.0)\n",
      "Requirement already satisfied: types-python-dateutil>=2.8.10 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from arrow>=0.15.0->isoduration->jsonschema==4.4.0->cas-tools) (2.9.0.20241206)\n",
      "Requirement already satisfied: jsonpath-ng in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from linkml-dataops->linkml==1.7.8->cas-tools) (1.7.0)\n",
      "Requirement already satisfied: ply in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from jsonpath-ng->linkml-dataops->linkml==1.7.8->cas-tools) (3.11)\n",
      "Requirement already satisfied: docopt>=0.6.2 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from num2words->quantulum3>=0.7.9->schema-automator==0.5.1->cas-tools) (0.6.2)\n",
      "Requirement already satisfied: sortedcontainers in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from pytrie->curies>=0.6.6->oaklib==0.6.4->cas-tools) (2.4.0)\n",
      "Requirement already satisfied: rdflib-jsonld==0.6.1 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from rdflib-shim->funowl>=0.2.0->oaklib==0.6.4->cas-tools) (0.6.1)\n",
      "\u001b[33mWARNING: scipy 1.13.1 does not provide the extra 'scipy'\u001b[0m\n",
      "Requirement already satisfied: stdlib_list in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from session-info->scanpy>=1.9.2->tiledbsoma~=1.11.4->cellxgene-census==1.15.0->cas-tools) (0.11.1)\n",
      "\u001b[33mWARNING: You are using pip version 21.2.4; however, version 25.1.1 is available.\n",
      "You should consider upgrading via the '/Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/bin/python3 -m pip install --upgrade pip' command.\u001b[0m\n",
      "Requirement already satisfied: anndata in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (0.10.5)\n",
      "Requirement already satisfied: scipy>1.4 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from anndata) (1.13.1)\n",
      "Requirement already satisfied: exceptiongroup in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from anndata) (1.2.2)\n",
      "Requirement already satisfied: pandas!=2.1.0rc0,!=2.1.2,>=1.1.1 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from anndata) (2.2.1)\n",
      "Requirement already satisfied: array-api-compat in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from anndata) (1.11.2)\n",
      "Requirement already satisfied: packaging>=20 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from anndata) (24.2)\n",
      "Requirement already satisfied: h5py>=3 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from anndata) (3.10.0)\n",
      "Requirement already satisfied: natsort in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from anndata) (8.4.0)\n",
      "Requirement already satisfied: numpy>=1.16.5 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from anndata) (1.26.4)\n",
      "Requirement already satisfied: pytz>=2020.1 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from pandas!=2.1.0rc0,!=2.1.2,>=1.1.1->anndata) (2025.1)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from pandas!=2.1.0rc0,!=2.1.2,>=1.1.1->anndata) (2.9.0)\n",
      "Requirement already satisfied: tzdata>=2022.7 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from pandas!=2.1.0rc0,!=2.1.2,>=1.1.1->anndata) (2025.1)\n",
      "Requirement already satisfied: six>=1.5 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from python-dateutil>=2.8.2->pandas!=2.1.0rc0,!=2.1.2,>=1.1.1->anndata) (1.17.0)\n",
      "\u001b[33mWARNING: You are using pip version 21.2.4; however, version 25.1.1 is available.\n",
      "You should consider upgrading via the '/Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/bin/python3 -m pip install --upgrade pip' command.\u001b[0m\n",
      "Requirement already satisfied: cachetools in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (5.5.2)\n",
      "\u001b[33mWARNING: You are using pip version 21.2.4; however, version 25.1.1 is available.\n",
      "You should consider upgrading via the '/Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/bin/python3 -m pip install --upgrade pip' command.\u001b[0m\n",
      "Requirement already satisfied: SPARQLWrapper in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (2.0.0)\n",
      "Requirement already satisfied: rdflib>=6.1.1 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from SPARQLWrapper) (7.0.0)\n",
      "Requirement already satisfied: pyparsing<4,>=2.1.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from rdflib>=6.1.1->SPARQLWrapper) (3.2.1)\n",
      "Requirement already satisfied: isodate<0.7.0,>=0.6.0 in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from rdflib>=6.1.1->SPARQLWrapper) (0.6.1)\n",
      "Requirement already satisfied: six in /Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/lib/python3.9/site-packages (from isodate<0.7.0,>=0.6.0->rdflib>=6.1.1->SPARQLWrapper) (1.17.0)\n",
      "\u001b[33mWARNING: You are using pip version 21.2.4; however, version 25.1.1 is available.\n",
      "You should consider upgrading via the '/Users/hk9/workspaces/workspace1/WMB_taxonomy_hacking/venv/bin/python3 -m pip install --upgrade pip' command.\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "# Install pip packages\n",
    "import sys\n",
    "!{sys.executable} -m pip install cas-tools\n",
    "!{sys.executable} -m pip install anndata\n",
    "!{sys.executable} -m pip install cachetools\n",
    "!{sys.executable} -m pip install SPARQLWrapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5a8844e6-f7ed-4543-ab9f-d370e95a8ccd",
   "metadata": {},
   "outputs": [],
   "source": [
    "## IMPORT depedencies \n",
    "\n",
    "import os\n",
    "import pandas as pd\n",
    "import anndata as ad\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7d5c0599-a0c5-45b1-830c-130e20474b9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "ann_data = ad.read_h5ad(\"/Users/hk9/tdt_datasets/Macaque_HMBA_basalganglia_AIT_pre-print.h5ad\", backed=\"r\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9a9344d2-145c-44b4-8ce9-7039f4b9527c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(548281, 103)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Neighborhood</th>\n",
       "      <th>Class</th>\n",
       "      <th>Subclass</th>\n",
       "      <th>Group</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>cell_type_ontology_term</th>\n",
       "      <th>load_id</th>\n",
       "      <th>donor_id</th>\n",
       "      <th>assay</th>\n",
       "      <th>assay_ontology_term_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>organism_ontology_term_id</th>\n",
       "      <th>development_stage</th>\n",
       "      <th>anatomical_region</th>\n",
       "      <th>anatomical_region_merged</th>\n",
       "      <th>anatomical_region_ontology_term_id</th>\n",
       "      <th>brain_region_ontology_term_id</th>\n",
       "      <th>self_reported_sex</th>\n",
       "      <th>self_reported_sex_ontology_term_id</th>\n",
       "      <th>self_reported_ethnicity</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>is_primary_data</th>\n",
       "      <th>atac_confidently_mapped_read_pairs</th>\n",
       "      <th>atac_fraction_of_genome_in_peaks</th>\n",
       "      <th>atac_fraction_of_high_quality_fragments_in_cells</th>\n",
       "      <th>atac_fraction_of_high_quality_fragments_overlapping_tss</th>\n",
       "      <th>atac_fraction_of_high_quality_fragments_overlapping_peaks</th>\n",
       "      <th>atac_fraction_of_transposition_events_in_peaks_in_cells</th>\n",
       "      <th>atac_mean_raw_read_pairs_per_cell</th>\n",
       "      <th>atac_median_high_quality_fragments_per_cell</th>\n",
       "      <th>atac_non_nuclear_read_pairs</th>\n",
       "      <th>atac_number_of_peaks</th>\n",
       "      <th>atac_percent_duplicates</th>\n",
       "      <th>atac_q30_bases_in_barcode</th>\n",
       "      <th>atac_q30_bases_in_read_1</th>\n",
       "      <th>atac_q30_bases_in_read_2</th>\n",
       "      <th>atac_q30_bases_in_sample_index_i1</th>\n",
       "      <th>atac_sequenced_read_pairs</th>\n",
       "      <th>atac_tss_enrichment_score</th>\n",
       "      <th>atac_unmapped_read_pairs</th>\n",
       "      <th>atac_valid_barcodes</th>\n",
       "      <th>gex_fraction_of_transcriptomic_reads_in_cells</th>\n",
       "      <th>gex_mean_raw_reads_per_cell</th>\n",
       "      <th>gex_median_umi_counts_per_cell</th>\n",
       "      <th>gex_median_genes_per_cell</th>\n",
       "      <th>gex_percent_duplicates</th>\n",
       "      <th>gex_q30_bases_in_umi</th>\n",
       "      <th>gex_q30_bases_in_barcode</th>\n",
       "      <th>gex_q30_bases_in_read_2</th>\n",
       "      <th>gex_q30_bases_in_sample_index_i1</th>\n",
       "      <th>gex_q30_bases_in_sample_index_i2</th>\n",
       "      <th>gex_reads_mapped_antisense_to_gene</th>\n",
       "      <th>gex_reads_mapped_confidently_to_exonic_regions</th>\n",
       "      <th>gex_reads_mapped_confidently_to_genome</th>\n",
       "      <th>gex_reads_mapped_confidently_to_intergenic_regions</th>\n",
       "      <th>gex_reads_mapped_confidently_to_intronic_regions</th>\n",
       "      <th>gex_reads_mapped_confidently_to_transcriptome</th>\n",
       "      <th>gex_reads_mapped_to_genome</th>\n",
       "      <th>gex_reads_with_tso</th>\n",
       "      <th>gex_sequenced_read_pairs</th>\n",
       "      <th>gex_total_genes_detected</th>\n",
       "      <th>gex_valid_umis</th>\n",
       "      <th>gex_valid_barcodes</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>total_genes</th>\n",
       "      <th>doublet_score</th>\n",
       "      <th>organism_sci</th>\n",
       "      <th>accession_group</th>\n",
       "      <th>embedding_set</th>\n",
       "      <th>CL:ID_group</th>\n",
       "      <th>tokens_group</th>\n",
       "      <th>display_order_group</th>\n",
       "      <th>literature_support</th>\n",
       "      <th>literature_name_short</th>\n",
       "      <th>literature_name_long</th>\n",
       "      <th>curated_markers</th>\n",
       "      <th>color_hex_group</th>\n",
       "      <th>color_group</th>\n",
       "      <th>accession_subclass</th>\n",
       "      <th>CL:ID_subclass</th>\n",
       "      <th>tokens_subclass</th>\n",
       "      <th>display_order_subclass</th>\n",
       "      <th>color_hex_subclass</th>\n",
       "      <th>color_subclass</th>\n",
       "      <th>accession_class</th>\n",
       "      <th>CL:ID_class</th>\n",
       "      <th>tokens_class</th>\n",
       "      <th>display_order_class</th>\n",
       "      <th>color_hex_class</th>\n",
       "      <th>color_class</th>\n",
       "      <th>accession_neighborhood</th>\n",
       "      <th>CL:ID_neighborhood</th>\n",
       "      <th>tokens_neighborhood</th>\n",
       "      <th>display_order_neighborhood</th>\n",
       "      <th>color_hex_neighborhood</th>\n",
       "      <th>color_neighborhood</th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>barcoded_cell_sample_label</th>\n",
       "      <th>alignment_job_database</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACAGCCAGTTGCGT-1061_B04</th>\n",
       "      <td>Subpallium GABA</td>\n",
       "      <td>CN LGE GABA</td>\n",
       "      <td>STR D1 MSN</td>\n",
       "      <td>STRv D1 MSN</td>\n",
       "      <td>Macaque-1</td>\n",
       "      <td>Macaque-1</td>\n",
       "      <td>CL:0000003</td>\n",
       "      <td>1061_B04</td>\n",
       "      <td>Q21.26.010</td>\n",
       "      <td>10x multiome</td>\n",
       "      <td>EFO_0030059</td>\n",
       "      <td>Macaque</td>\n",
       "      <td>NCBITaxon:9544</td>\n",
       "      <td>Adult</td>\n",
       "      <td>NAC</td>\n",
       "      <td>NAC</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Male</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Macaca Mulatta</td>\n",
       "      <td>normal</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>True</td>\n",
       "      <td>0.8805</td>\n",
       "      <td>0.028</td>\n",
       "      <td>0.7524</td>\n",
       "      <td>0.213</td>\n",
       "      <td>0.2391</td>\n",
       "      <td>0.2184</td>\n",
       "      <td>95529.2519</td>\n",
       "      <td>35407.0</td>\n",
       "      <td>0</td>\n",
       "      <td>95381</td>\n",
       "      <td>0.3599</td>\n",
       "      <td>0.9235</td>\n",
       "      <td>0.9632</td>\n",
       "      <td>0.9532</td>\n",
       "      <td>0.9523</td>\n",
       "      <td>525315356</td>\n",
       "      <td>4.6613</td>\n",
       "      <td>0.0107</td>\n",
       "      <td>0.9846</td>\n",
       "      <td>0.9301</td>\n",
       "      <td>42172.1289</td>\n",
       "      <td>10016.0</td>\n",
       "      <td>3875.0</td>\n",
       "      <td>0.3316</td>\n",
       "      <td>0.9387</td>\n",
       "      <td>0.9398</td>\n",
       "      <td>0.9272</td>\n",
       "      <td>0.9373</td>\n",
       "      <td>0.9069</td>\n",
       "      <td>0.341</td>\n",
       "      <td>0.1172</td>\n",
       "      <td>0.9076</td>\n",
       "      <td>0.1177</td>\n",
       "      <td>0.6727</td>\n",
       "      <td>0.4463</td>\n",
       "      <td>0.9485</td>\n",
       "      <td>0.1083</td>\n",
       "      <td>231904537</td>\n",
       "      <td>29345</td>\n",
       "      <td>0.9994</td>\n",
       "      <td>0.892</td>\n",
       "      <td>14195</td>\n",
       "      <td>4639</td>\n",
       "      <td>0.018182</td>\n",
       "      <td>Macaca Mulatta</td>\n",
       "      <td>CS20250428_GROUP_0056</td>\n",
       "      <td>LGE</td>\n",
       "      <td>CL:4030051</td>\n",
       "      <td>STRv|D1 MSN</td>\n",
       "      <td>51</td>\n",
       "      <td>He et al. 2021</td>\n",
       "      <td>D1-Shell/OT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ARHGAP6</td>\n",
       "      <td>#339933</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CS20250428_SUBCL_0028</td>\n",
       "      <td>CL:4023026</td>\n",
       "      <td>STR|D1 MSN</td>\n",
       "      <td>31</td>\n",
       "      <td>#1655f2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CS20250428_CLASS_0003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CN|LGE|GABA</td>\n",
       "      <td>10</td>\n",
       "      <td>#d0b83c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CS20250428_NEIGH_0002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Subpallium|GABA</td>\n",
       "      <td>3</td>\n",
       "      <td>#19613b</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AAACAGCCAGTTGCGT</td>\n",
       "      <td>1061_B04</td>\n",
       "      <td>AIBS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATGCAATTGACT-1061_B04</th>\n",
       "      <td>Subpallium GABA</td>\n",
       "      <td>F M GABA</td>\n",
       "      <td>F GABA</td>\n",
       "      <td>AMY-SLEA-BNST GABA</td>\n",
       "      <td>Macaque-18</td>\n",
       "      <td>Macaque-18</td>\n",
       "      <td>CL:0000003</td>\n",
       "      <td>1061_B04</td>\n",
       "      <td>Q21.26.010</td>\n",
       "      <td>10x multiome</td>\n",
       "      <td>EFO_0030059</td>\n",
       "      <td>Macaque</td>\n",
       "      <td>NCBITaxon:9544</td>\n",
       "      <td>Adult</td>\n",
       "      <td>NAC</td>\n",
       "      <td>NAC</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Male</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Macaca Mulatta</td>\n",
       "      <td>normal</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>True</td>\n",
       "      <td>0.8805</td>\n",
       "      <td>0.028</td>\n",
       "      <td>0.7524</td>\n",
       "      <td>0.213</td>\n",
       "      <td>0.2391</td>\n",
       "      <td>0.2184</td>\n",
       "      <td>95529.2519</td>\n",
       "      <td>35407.0</td>\n",
       "      <td>0</td>\n",
       "      <td>95381</td>\n",
       "      <td>0.3599</td>\n",
       "      <td>0.9235</td>\n",
       "      <td>0.9632</td>\n",
       "      <td>0.9532</td>\n",
       "      <td>0.9523</td>\n",
       "      <td>525315356</td>\n",
       "      <td>4.6613</td>\n",
       "      <td>0.0107</td>\n",
       "      <td>0.9846</td>\n",
       "      <td>0.9301</td>\n",
       "      <td>42172.1289</td>\n",
       "      <td>10016.0</td>\n",
       "      <td>3875.0</td>\n",
       "      <td>0.3316</td>\n",
       "      <td>0.9387</td>\n",
       "      <td>0.9398</td>\n",
       "      <td>0.9272</td>\n",
       "      <td>0.9373</td>\n",
       "      <td>0.9069</td>\n",
       "      <td>0.341</td>\n",
       "      <td>0.1172</td>\n",
       "      <td>0.9076</td>\n",
       "      <td>0.1177</td>\n",
       "      <td>0.6727</td>\n",
       "      <td>0.4463</td>\n",
       "      <td>0.9485</td>\n",
       "      <td>0.1083</td>\n",
       "      <td>231904537</td>\n",
       "      <td>29345</td>\n",
       "      <td>0.9994</td>\n",
       "      <td>0.892</td>\n",
       "      <td>4441</td>\n",
       "      <td>2507</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Macaca Mulatta</td>\n",
       "      <td>CS20250428_GROUP_0066</td>\n",
       "      <td>Interneuron</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AMY|SLEA|BNST|GABA</td>\n",
       "      <td>48</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>#b69e14</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CS20250428_SUBCL_0037</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F|GABA</td>\n",
       "      <td>29</td>\n",
       "      <td>#7ba2df</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CS20250428_CLASS_0006</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F|M|GABA</td>\n",
       "      <td>9</td>\n",
       "      <td>#cd0f13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CS20250428_NEIGH_0002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Subpallium|GABA</td>\n",
       "      <td>3</td>\n",
       "      <td>#19613b</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AAACATGCAATTGACT</td>\n",
       "      <td>1061_B04</td>\n",
       "      <td>AIBS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATGCACGTTACA-1061_B04</th>\n",
       "      <td>Subpallium GABA</td>\n",
       "      <td>CN LGE GABA</td>\n",
       "      <td>STR Hybrid MSN</td>\n",
       "      <td>STRv D1 NUDAP MSN</td>\n",
       "      <td>Macaque-35</td>\n",
       "      <td>Macaque-35</td>\n",
       "      <td>CL:0000003</td>\n",
       "      <td>1061_B04</td>\n",
       "      <td>Q21.26.010</td>\n",
       "      <td>10x multiome</td>\n",
       "      <td>EFO_0030059</td>\n",
       "      <td>Macaque</td>\n",
       "      <td>NCBITaxon:9544</td>\n",
       "      <td>Adult</td>\n",
       "      <td>NAC</td>\n",
       "      <td>NAC</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Male</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Macaca Mulatta</td>\n",
       "      <td>normal</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>True</td>\n",
       "      <td>0.8805</td>\n",
       "      <td>0.028</td>\n",
       "      <td>0.7524</td>\n",
       "      <td>0.213</td>\n",
       "      <td>0.2391</td>\n",
       "      <td>0.2184</td>\n",
       "      <td>95529.2519</td>\n",
       "      <td>35407.0</td>\n",
       "      <td>0</td>\n",
       "      <td>95381</td>\n",
       "      <td>0.3599</td>\n",
       "      <td>0.9235</td>\n",
       "      <td>0.9632</td>\n",
       "      <td>0.9532</td>\n",
       "      <td>0.9523</td>\n",
       "      <td>525315356</td>\n",
       "      <td>4.6613</td>\n",
       "      <td>0.0107</td>\n",
       "      <td>0.9846</td>\n",
       "      <td>0.9301</td>\n",
       "      <td>42172.1289</td>\n",
       "      <td>10016.0</td>\n",
       "      <td>3875.0</td>\n",
       "      <td>0.3316</td>\n",
       "      <td>0.9387</td>\n",
       "      <td>0.9398</td>\n",
       "      <td>0.9272</td>\n",
       "      <td>0.9373</td>\n",
       "      <td>0.9069</td>\n",
       "      <td>0.341</td>\n",
       "      <td>0.1172</td>\n",
       "      <td>0.9076</td>\n",
       "      <td>0.1177</td>\n",
       "      <td>0.6727</td>\n",
       "      <td>0.4463</td>\n",
       "      <td>0.9485</td>\n",
       "      <td>0.1083</td>\n",
       "      <td>231904537</td>\n",
       "      <td>29345</td>\n",
       "      <td>0.9994</td>\n",
       "      <td>0.892</td>\n",
       "      <td>10473</td>\n",
       "      <td>3808</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Macaca Mulatta</td>\n",
       "      <td>CS20250428_GROUP_0057</td>\n",
       "      <td>LGE</td>\n",
       "      <td>CL:4030054</td>\n",
       "      <td>STRv|D1 NUDAP MSN</td>\n",
       "      <td>57</td>\n",
       "      <td>He et al. 2021</td>\n",
       "      <td>D1-NUDAP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GRIA4</td>\n",
       "      <td>#e377c2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CS20250428_SUBCL_0030</td>\n",
       "      <td>CL:4030050</td>\n",
       "      <td>STR|Hybrid MSN</td>\n",
       "      <td>33</td>\n",
       "      <td>#79bdf4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CS20250428_CLASS_0003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CN|LGE|GABA</td>\n",
       "      <td>10</td>\n",
       "      <td>#d0b83c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CS20250428_NEIGH_0002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Subpallium|GABA</td>\n",
       "      <td>3</td>\n",
       "      <td>#19613b</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AAACATGCACGTTACA</td>\n",
       "      <td>1061_B04</td>\n",
       "      <td>AIBS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATGCAGGTTACC-1061_B04</th>\n",
       "      <td>Subpallium GABA</td>\n",
       "      <td>CN LGE GABA</td>\n",
       "      <td>STR Hybrid MSN</td>\n",
       "      <td>STRv D1 NUDAP MSN</td>\n",
       "      <td>Macaque-31</td>\n",
       "      <td>Macaque-31</td>\n",
       "      <td>CL:0000003</td>\n",
       "      <td>1061_B04</td>\n",
       "      <td>Q21.26.010</td>\n",
       "      <td>10x multiome</td>\n",
       "      <td>EFO_0030059</td>\n",
       "      <td>Macaque</td>\n",
       "      <td>NCBITaxon:9544</td>\n",
       "      <td>Adult</td>\n",
       "      <td>NAC</td>\n",
       "      <td>NAC</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Male</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Macaca Mulatta</td>\n",
       "      <td>normal</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>True</td>\n",
       "      <td>0.8805</td>\n",
       "      <td>0.028</td>\n",
       "      <td>0.7524</td>\n",
       "      <td>0.213</td>\n",
       "      <td>0.2391</td>\n",
       "      <td>0.2184</td>\n",
       "      <td>95529.2519</td>\n",
       "      <td>35407.0</td>\n",
       "      <td>0</td>\n",
       "      <td>95381</td>\n",
       "      <td>0.3599</td>\n",
       "      <td>0.9235</td>\n",
       "      <td>0.9632</td>\n",
       "      <td>0.9532</td>\n",
       "      <td>0.9523</td>\n",
       "      <td>525315356</td>\n",
       "      <td>4.6613</td>\n",
       "      <td>0.0107</td>\n",
       "      <td>0.9846</td>\n",
       "      <td>0.9301</td>\n",
       "      <td>42172.1289</td>\n",
       "      <td>10016.0</td>\n",
       "      <td>3875.0</td>\n",
       "      <td>0.3316</td>\n",
       "      <td>0.9387</td>\n",
       "      <td>0.9398</td>\n",
       "      <td>0.9272</td>\n",
       "      <td>0.9373</td>\n",
       "      <td>0.9069</td>\n",
       "      <td>0.341</td>\n",
       "      <td>0.1172</td>\n",
       "      <td>0.9076</td>\n",
       "      <td>0.1177</td>\n",
       "      <td>0.6727</td>\n",
       "      <td>0.4463</td>\n",
       "      <td>0.9485</td>\n",
       "      <td>0.1083</td>\n",
       "      <td>231904537</td>\n",
       "      <td>29345</td>\n",
       "      <td>0.9994</td>\n",
       "      <td>0.892</td>\n",
       "      <td>11817</td>\n",
       "      <td>4080</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Macaca Mulatta</td>\n",
       "      <td>CS20250428_GROUP_0057</td>\n",
       "      <td>LGE</td>\n",
       "      <td>CL:4030054</td>\n",
       "      <td>STRv|D1 NUDAP MSN</td>\n",
       "      <td>57</td>\n",
       "      <td>He et al. 2021</td>\n",
       "      <td>D1-NUDAP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GRIA4</td>\n",
       "      <td>#e377c2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CS20250428_SUBCL_0030</td>\n",
       "      <td>CL:4030050</td>\n",
       "      <td>STR|Hybrid MSN</td>\n",
       "      <td>33</td>\n",
       "      <td>#79bdf4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CS20250428_CLASS_0003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CN|LGE|GABA</td>\n",
       "      <td>10</td>\n",
       "      <td>#d0b83c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CS20250428_NEIGH_0002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Subpallium|GABA</td>\n",
       "      <td>3</td>\n",
       "      <td>#19613b</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AAACATGCAGGTTACC</td>\n",
       "      <td>1061_B04</td>\n",
       "      <td>AIBS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              Neighborhood        Class        Subclass  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04  Subpallium GABA  CN LGE GABA      STR D1 MSN   \n",
       "AAACATGCAATTGACT-1061_B04  Subpallium GABA     F M GABA          F GABA   \n",
       "AAACATGCACGTTACA-1061_B04  Subpallium GABA  CN LGE GABA  STR Hybrid MSN   \n",
       "AAACATGCAGGTTACC-1061_B04  Subpallium GABA  CN LGE GABA  STR Hybrid MSN   \n",
       "\n",
       "                                        Group     Cluster  cluster_id  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04         STRv D1 MSN   Macaque-1   Macaque-1   \n",
       "AAACATGCAATTGACT-1061_B04  AMY-SLEA-BNST GABA  Macaque-18  Macaque-18   \n",
       "AAACATGCACGTTACA-1061_B04   STRv D1 NUDAP MSN  Macaque-35  Macaque-35   \n",
       "AAACATGCAGGTTACC-1061_B04   STRv D1 NUDAP MSN  Macaque-31  Macaque-31   \n",
       "\n",
       "                          cell_type_ontology_term   load_id    donor_id  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04              CL:0000003  1061_B04  Q21.26.010   \n",
       "AAACATGCAATTGACT-1061_B04              CL:0000003  1061_B04  Q21.26.010   \n",
       "AAACATGCACGTTACA-1061_B04              CL:0000003  1061_B04  Q21.26.010   \n",
       "AAACATGCAGGTTACC-1061_B04              CL:0000003  1061_B04  Q21.26.010   \n",
       "\n",
       "                                  assay assay_ontology_term_id organism  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04  10x multiome            EFO_0030059  Macaque   \n",
       "AAACATGCAATTGACT-1061_B04  10x multiome            EFO_0030059  Macaque   \n",
       "AAACATGCACGTTACA-1061_B04  10x multiome            EFO_0030059  Macaque   \n",
       "AAACATGCAGGTTACC-1061_B04  10x multiome            EFO_0030059  Macaque   \n",
       "\n",
       "                          organism_ontology_term_id development_stage  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04            NCBITaxon:9544             Adult   \n",
       "AAACATGCAATTGACT-1061_B04            NCBITaxon:9544             Adult   \n",
       "AAACATGCACGTTACA-1061_B04            NCBITaxon:9544             Adult   \n",
       "AAACATGCAGGTTACC-1061_B04            NCBITaxon:9544             Adult   \n",
       "\n",
       "                          anatomical_region anatomical_region_merged  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04               NAC                      NAC   \n",
       "AAACATGCAATTGACT-1061_B04               NAC                      NAC   \n",
       "AAACATGCACGTTACA-1061_B04               NAC                      NAC   \n",
       "AAACATGCAGGTTACC-1061_B04               NAC                      NAC   \n",
       "\n",
       "                          anatomical_region_ontology_term_id  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                            Unknown   \n",
       "AAACATGCAATTGACT-1061_B04                            Unknown   \n",
       "AAACATGCACGTTACA-1061_B04                            Unknown   \n",
       "AAACATGCAGGTTACC-1061_B04                            Unknown   \n",
       "\n",
       "                          brain_region_ontology_term_id self_reported_sex  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                       Unknown              Male   \n",
       "AAACATGCAATTGACT-1061_B04                       Unknown              Male   \n",
       "AAACATGCACGTTACA-1061_B04                       Unknown              Male   \n",
       "AAACATGCAGGTTACC-1061_B04                       Unknown              Male   \n",
       "\n",
       "                          self_reported_sex_ontology_term_id  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                       PATO:0000384   \n",
       "AAACATGCAATTGACT-1061_B04                       PATO:0000384   \n",
       "AAACATGCACGTTACA-1061_B04                       PATO:0000384   \n",
       "AAACATGCAGGTTACC-1061_B04                       PATO:0000384   \n",
       "\n",
       "                          self_reported_ethnicity  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                 unknown   \n",
       "AAACATGCAATTGACT-1061_B04                 unknown   \n",
       "AAACATGCACGTTACA-1061_B04                 unknown   \n",
       "AAACATGCAGGTTACC-1061_B04                 unknown   \n",
       "\n",
       "                          self_reported_ethnicity_ontology_term_id disease  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                           Macaca Mulatta  normal   \n",
       "AAACATGCAATTGACT-1061_B04                           Macaca Mulatta  normal   \n",
       "AAACATGCACGTTACA-1061_B04                           Macaca Mulatta  normal   \n",
       "AAACATGCAGGTTACC-1061_B04                           Macaca Mulatta  normal   \n",
       "\n",
       "                          disease_ontology_term_id suspension_type  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04             PATO:0000461         nucleus   \n",
       "AAACATGCAATTGACT-1061_B04             PATO:0000461         nucleus   \n",
       "AAACATGCACGTTACA-1061_B04             PATO:0000461         nucleus   \n",
       "AAACATGCAGGTTACC-1061_B04             PATO:0000461         nucleus   \n",
       "\n",
       "                          is_primary_data  atac_confidently_mapped_read_pairs  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04            True                              0.8805   \n",
       "AAACATGCAATTGACT-1061_B04            True                              0.8805   \n",
       "AAACATGCACGTTACA-1061_B04            True                              0.8805   \n",
       "AAACATGCAGGTTACC-1061_B04            True                              0.8805   \n",
       "\n",
       "                           atac_fraction_of_genome_in_peaks  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                             0.028   \n",
       "AAACATGCAATTGACT-1061_B04                             0.028   \n",
       "AAACATGCACGTTACA-1061_B04                             0.028   \n",
       "AAACATGCAGGTTACC-1061_B04                             0.028   \n",
       "\n",
       "                           atac_fraction_of_high_quality_fragments_in_cells  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                                            0.7524   \n",
       "AAACATGCAATTGACT-1061_B04                                            0.7524   \n",
       "AAACATGCACGTTACA-1061_B04                                            0.7524   \n",
       "AAACATGCAGGTTACC-1061_B04                                            0.7524   \n",
       "\n",
       "                           atac_fraction_of_high_quality_fragments_overlapping_tss  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                                              0.213         \n",
       "AAACATGCAATTGACT-1061_B04                                              0.213         \n",
       "AAACATGCACGTTACA-1061_B04                                              0.213         \n",
       "AAACATGCAGGTTACC-1061_B04                                              0.213         \n",
       "\n",
       "                           atac_fraction_of_high_quality_fragments_overlapping_peaks  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                                             0.2391           \n",
       "AAACATGCAATTGACT-1061_B04                                             0.2391           \n",
       "AAACATGCACGTTACA-1061_B04                                             0.2391           \n",
       "AAACATGCAGGTTACC-1061_B04                                             0.2391           \n",
       "\n",
       "                           atac_fraction_of_transposition_events_in_peaks_in_cells  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                                             0.2184         \n",
       "AAACATGCAATTGACT-1061_B04                                             0.2184         \n",
       "AAACATGCACGTTACA-1061_B04                                             0.2184         \n",
       "AAACATGCAGGTTACC-1061_B04                                             0.2184         \n",
       "\n",
       "                           atac_mean_raw_read_pairs_per_cell  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                         95529.2519   \n",
       "AAACATGCAATTGACT-1061_B04                         95529.2519   \n",
       "AAACATGCACGTTACA-1061_B04                         95529.2519   \n",
       "AAACATGCAGGTTACC-1061_B04                         95529.2519   \n",
       "\n",
       "                           atac_median_high_quality_fragments_per_cell  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                                      35407.0   \n",
       "AAACATGCAATTGACT-1061_B04                                      35407.0   \n",
       "AAACATGCACGTTACA-1061_B04                                      35407.0   \n",
       "AAACATGCAGGTTACC-1061_B04                                      35407.0   \n",
       "\n",
       "                           atac_non_nuclear_read_pairs  atac_number_of_peaks  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                            0                 95381   \n",
       "AAACATGCAATTGACT-1061_B04                            0                 95381   \n",
       "AAACATGCACGTTACA-1061_B04                            0                 95381   \n",
       "AAACATGCAGGTTACC-1061_B04                            0                 95381   \n",
       "\n",
       "                           atac_percent_duplicates  atac_q30_bases_in_barcode  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                   0.3599                     0.9235   \n",
       "AAACATGCAATTGACT-1061_B04                   0.3599                     0.9235   \n",
       "AAACATGCACGTTACA-1061_B04                   0.3599                     0.9235   \n",
       "AAACATGCAGGTTACC-1061_B04                   0.3599                     0.9235   \n",
       "\n",
       "                           atac_q30_bases_in_read_1  atac_q30_bases_in_read_2  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                    0.9632                    0.9532   \n",
       "AAACATGCAATTGACT-1061_B04                    0.9632                    0.9532   \n",
       "AAACATGCACGTTACA-1061_B04                    0.9632                    0.9532   \n",
       "AAACATGCAGGTTACC-1061_B04                    0.9632                    0.9532   \n",
       "\n",
       "                           atac_q30_bases_in_sample_index_i1  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                             0.9523   \n",
       "AAACATGCAATTGACT-1061_B04                             0.9523   \n",
       "AAACATGCACGTTACA-1061_B04                             0.9523   \n",
       "AAACATGCAGGTTACC-1061_B04                             0.9523   \n",
       "\n",
       "                           atac_sequenced_read_pairs  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                  525315356   \n",
       "AAACATGCAATTGACT-1061_B04                  525315356   \n",
       "AAACATGCACGTTACA-1061_B04                  525315356   \n",
       "AAACATGCAGGTTACC-1061_B04                  525315356   \n",
       "\n",
       "                           atac_tss_enrichment_score  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                     4.6613   \n",
       "AAACATGCAATTGACT-1061_B04                     4.6613   \n",
       "AAACATGCACGTTACA-1061_B04                     4.6613   \n",
       "AAACATGCAGGTTACC-1061_B04                     4.6613   \n",
       "\n",
       "                           atac_unmapped_read_pairs  atac_valid_barcodes  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                    0.0107               0.9846   \n",
       "AAACATGCAATTGACT-1061_B04                    0.0107               0.9846   \n",
       "AAACATGCACGTTACA-1061_B04                    0.0107               0.9846   \n",
       "AAACATGCAGGTTACC-1061_B04                    0.0107               0.9846   \n",
       "\n",
       "                           gex_fraction_of_transcriptomic_reads_in_cells  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                                         0.9301   \n",
       "AAACATGCAATTGACT-1061_B04                                         0.9301   \n",
       "AAACATGCACGTTACA-1061_B04                                         0.9301   \n",
       "AAACATGCAGGTTACC-1061_B04                                         0.9301   \n",
       "\n",
       "                           gex_mean_raw_reads_per_cell  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                   42172.1289   \n",
       "AAACATGCAATTGACT-1061_B04                   42172.1289   \n",
       "AAACATGCACGTTACA-1061_B04                   42172.1289   \n",
       "AAACATGCAGGTTACC-1061_B04                   42172.1289   \n",
       "\n",
       "                           gex_median_umi_counts_per_cell  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                         10016.0   \n",
       "AAACATGCAATTGACT-1061_B04                         10016.0   \n",
       "AAACATGCACGTTACA-1061_B04                         10016.0   \n",
       "AAACATGCAGGTTACC-1061_B04                         10016.0   \n",
       "\n",
       "                           gex_median_genes_per_cell  gex_percent_duplicates  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                     3875.0                  0.3316   \n",
       "AAACATGCAATTGACT-1061_B04                     3875.0                  0.3316   \n",
       "AAACATGCACGTTACA-1061_B04                     3875.0                  0.3316   \n",
       "AAACATGCAGGTTACC-1061_B04                     3875.0                  0.3316   \n",
       "\n",
       "                           gex_q30_bases_in_umi  gex_q30_bases_in_barcode  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                0.9387                    0.9398   \n",
       "AAACATGCAATTGACT-1061_B04                0.9387                    0.9398   \n",
       "AAACATGCACGTTACA-1061_B04                0.9387                    0.9398   \n",
       "AAACATGCAGGTTACC-1061_B04                0.9387                    0.9398   \n",
       "\n",
       "                           gex_q30_bases_in_read_2  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                   0.9272   \n",
       "AAACATGCAATTGACT-1061_B04                   0.9272   \n",
       "AAACATGCACGTTACA-1061_B04                   0.9272   \n",
       "AAACATGCAGGTTACC-1061_B04                   0.9272   \n",
       "\n",
       "                           gex_q30_bases_in_sample_index_i1  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                            0.9373   \n",
       "AAACATGCAATTGACT-1061_B04                            0.9373   \n",
       "AAACATGCACGTTACA-1061_B04                            0.9373   \n",
       "AAACATGCAGGTTACC-1061_B04                            0.9373   \n",
       "\n",
       "                           gex_q30_bases_in_sample_index_i2  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                            0.9069   \n",
       "AAACATGCAATTGACT-1061_B04                            0.9069   \n",
       "AAACATGCACGTTACA-1061_B04                            0.9069   \n",
       "AAACATGCAGGTTACC-1061_B04                            0.9069   \n",
       "\n",
       "                           gex_reads_mapped_antisense_to_gene  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                               0.341   \n",
       "AAACATGCAATTGACT-1061_B04                               0.341   \n",
       "AAACATGCACGTTACA-1061_B04                               0.341   \n",
       "AAACATGCAGGTTACC-1061_B04                               0.341   \n",
       "\n",
       "                           gex_reads_mapped_confidently_to_exonic_regions  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                                          0.1172   \n",
       "AAACATGCAATTGACT-1061_B04                                          0.1172   \n",
       "AAACATGCACGTTACA-1061_B04                                          0.1172   \n",
       "AAACATGCAGGTTACC-1061_B04                                          0.1172   \n",
       "\n",
       "                           gex_reads_mapped_confidently_to_genome  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                                  0.9076   \n",
       "AAACATGCAATTGACT-1061_B04                                  0.9076   \n",
       "AAACATGCACGTTACA-1061_B04                                  0.9076   \n",
       "AAACATGCAGGTTACC-1061_B04                                  0.9076   \n",
       "\n",
       "                           gex_reads_mapped_confidently_to_intergenic_regions  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                                             0.1177    \n",
       "AAACATGCAATTGACT-1061_B04                                             0.1177    \n",
       "AAACATGCACGTTACA-1061_B04                                             0.1177    \n",
       "AAACATGCAGGTTACC-1061_B04                                             0.1177    \n",
       "\n",
       "                           gex_reads_mapped_confidently_to_intronic_regions  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                                            0.6727   \n",
       "AAACATGCAATTGACT-1061_B04                                            0.6727   \n",
       "AAACATGCACGTTACA-1061_B04                                            0.6727   \n",
       "AAACATGCAGGTTACC-1061_B04                                            0.6727   \n",
       "\n",
       "                           gex_reads_mapped_confidently_to_transcriptome  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                                         0.4463   \n",
       "AAACATGCAATTGACT-1061_B04                                         0.4463   \n",
       "AAACATGCACGTTACA-1061_B04                                         0.4463   \n",
       "AAACATGCAGGTTACC-1061_B04                                         0.4463   \n",
       "\n",
       "                           gex_reads_mapped_to_genome  gex_reads_with_tso  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                      0.9485              0.1083   \n",
       "AAACATGCAATTGACT-1061_B04                      0.9485              0.1083   \n",
       "AAACATGCACGTTACA-1061_B04                      0.9485              0.1083   \n",
       "AAACATGCAGGTTACC-1061_B04                      0.9485              0.1083   \n",
       "\n",
       "                           gex_sequenced_read_pairs  gex_total_genes_detected  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                 231904537                     29345   \n",
       "AAACATGCAATTGACT-1061_B04                 231904537                     29345   \n",
       "AAACATGCACGTTACA-1061_B04                 231904537                     29345   \n",
       "AAACATGCAGGTTACC-1061_B04                 231904537                     29345   \n",
       "\n",
       "                           gex_valid_umis  gex_valid_barcodes  total_counts  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04          0.9994               0.892         14195   \n",
       "AAACATGCAATTGACT-1061_B04          0.9994               0.892          4441   \n",
       "AAACATGCACGTTACA-1061_B04          0.9994               0.892         10473   \n",
       "AAACATGCAGGTTACC-1061_B04          0.9994               0.892         11817   \n",
       "\n",
       "                           total_genes  doublet_score    organism_sci  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04         4639       0.018182  Macaca Mulatta   \n",
       "AAACATGCAATTGACT-1061_B04         2507       0.000000  Macaca Mulatta   \n",
       "AAACATGCACGTTACA-1061_B04         3808       0.000000  Macaca Mulatta   \n",
       "AAACATGCAGGTTACC-1061_B04         4080       0.000000  Macaca Mulatta   \n",
       "\n",
       "                                 accession_group embedding_set CL:ID_group  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04  CS20250428_GROUP_0056           LGE  CL:4030051   \n",
       "AAACATGCAATTGACT-1061_B04  CS20250428_GROUP_0066   Interneuron         NaN   \n",
       "AAACATGCACGTTACA-1061_B04  CS20250428_GROUP_0057           LGE  CL:4030054   \n",
       "AAACATGCAGGTTACC-1061_B04  CS20250428_GROUP_0057           LGE  CL:4030054   \n",
       "\n",
       "                                 tokens_group  display_order_group  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04         STRv|D1 MSN                   51   \n",
       "AAACATGCAATTGACT-1061_B04  AMY|SLEA|BNST|GABA                   48   \n",
       "AAACATGCACGTTACA-1061_B04   STRv|D1 NUDAP MSN                   57   \n",
       "AAACATGCAGGTTACC-1061_B04   STRv|D1 NUDAP MSN                   57   \n",
       "\n",
       "                          literature_support literature_name_short  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04     He et al. 2021           D1-Shell/OT   \n",
       "AAACATGCAATTGACT-1061_B04                NaN                   NaN   \n",
       "AAACATGCACGTTACA-1061_B04     He et al. 2021              D1-NUDAP   \n",
       "AAACATGCAGGTTACC-1061_B04     He et al. 2021              D1-NUDAP   \n",
       "\n",
       "                          literature_name_long curated_markers  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                  NaN         ARHGAP6   \n",
       "AAACATGCAATTGACT-1061_B04                  NaN             NaN   \n",
       "AAACATGCACGTTACA-1061_B04                  NaN           GRIA4   \n",
       "AAACATGCAGGTTACC-1061_B04                  NaN           GRIA4   \n",
       "\n",
       "                          color_hex_group  color_group     accession_subclass  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04         #339933          NaN  CS20250428_SUBCL_0028   \n",
       "AAACATGCAATTGACT-1061_B04         #b69e14          NaN  CS20250428_SUBCL_0037   \n",
       "AAACATGCACGTTACA-1061_B04         #e377c2          NaN  CS20250428_SUBCL_0030   \n",
       "AAACATGCAGGTTACC-1061_B04         #e377c2          NaN  CS20250428_SUBCL_0030   \n",
       "\n",
       "                          CL:ID_subclass tokens_subclass  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04     CL:4023026      STR|D1 MSN   \n",
       "AAACATGCAATTGACT-1061_B04            NaN          F|GABA   \n",
       "AAACATGCACGTTACA-1061_B04     CL:4030050  STR|Hybrid MSN   \n",
       "AAACATGCAGGTTACC-1061_B04     CL:4030050  STR|Hybrid MSN   \n",
       "\n",
       "                           display_order_subclass color_hex_subclass  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                      31            #1655f2   \n",
       "AAACATGCAATTGACT-1061_B04                      29            #7ba2df   \n",
       "AAACATGCACGTTACA-1061_B04                      33            #79bdf4   \n",
       "AAACATGCAGGTTACC-1061_B04                      33            #79bdf4   \n",
       "\n",
       "                           color_subclass        accession_class CL:ID_class  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04             NaN  CS20250428_CLASS_0003         NaN   \n",
       "AAACATGCAATTGACT-1061_B04             NaN  CS20250428_CLASS_0006         NaN   \n",
       "AAACATGCACGTTACA-1061_B04             NaN  CS20250428_CLASS_0003         NaN   \n",
       "AAACATGCAGGTTACC-1061_B04             NaN  CS20250428_CLASS_0003         NaN   \n",
       "\n",
       "                          tokens_class  display_order_class color_hex_class  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04  CN|LGE|GABA                   10         #d0b83c   \n",
       "AAACATGCAATTGACT-1061_B04     F|M|GABA                    9         #cd0f13   \n",
       "AAACATGCACGTTACA-1061_B04  CN|LGE|GABA                   10         #d0b83c   \n",
       "AAACATGCAGGTTACC-1061_B04  CN|LGE|GABA                   10         #d0b83c   \n",
       "\n",
       "                           color_class accession_neighborhood  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04          NaN  CS20250428_NEIGH_0002   \n",
       "AAACATGCAATTGACT-1061_B04          NaN  CS20250428_NEIGH_0002   \n",
       "AAACATGCACGTTACA-1061_B04          NaN  CS20250428_NEIGH_0002   \n",
       "AAACATGCAGGTTACC-1061_B04          NaN  CS20250428_NEIGH_0002   \n",
       "\n",
       "                          CL:ID_neighborhood tokens_neighborhood  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                NaN     Subpallium|GABA   \n",
       "AAACATGCAATTGACT-1061_B04                NaN     Subpallium|GABA   \n",
       "AAACATGCACGTTACA-1061_B04                NaN     Subpallium|GABA   \n",
       "AAACATGCAGGTTACC-1061_B04                NaN     Subpallium|GABA   \n",
       "\n",
       "                           display_order_neighborhood color_hex_neighborhood  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                           3                #19613b   \n",
       "AAACATGCAATTGACT-1061_B04                           3                #19613b   \n",
       "AAACATGCACGTTACA-1061_B04                           3                #19613b   \n",
       "AAACATGCAGGTTACC-1061_B04                           3                #19613b   \n",
       "\n",
       "                           color_neighborhood      cell_barcode  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                 NaN  AAACAGCCAGTTGCGT   \n",
       "AAACATGCAATTGACT-1061_B04                 NaN  AAACATGCAATTGACT   \n",
       "AAACATGCACGTTACA-1061_B04                 NaN  AAACATGCACGTTACA   \n",
       "AAACATGCAGGTTACC-1061_B04                 NaN  AAACATGCAGGTTACC   \n",
       "\n",
       "                          barcoded_cell_sample_label alignment_job_database  \n",
       "AAACAGCCAGTTGCGT-1061_B04                   1061_B04                   AIBS  \n",
       "AAACATGCAATTGACT-1061_B04                   1061_B04                   AIBS  \n",
       "AAACATGCACGTTACA-1061_B04                   1061_B04                   AIBS  \n",
       "AAACATGCAGGTTACC-1061_B04                   1061_B04                   AIBS  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', None)\n",
    "print(ann_data.obs.shape)\n",
    "ann_data.obs.head(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "df544ee7-7a1c-42a6-a51b-8f913129a195",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Neighborhood',\n",
       " 'Class',\n",
       " 'Subclass',\n",
       " 'Group',\n",
       " 'Cluster',\n",
       " 'cluster_id',\n",
       " 'cell_type_ontology_term',\n",
       " 'load_id',\n",
       " 'donor_id',\n",
       " 'assay',\n",
       " 'assay_ontology_term_id',\n",
       " 'organism',\n",
       " 'organism_ontology_term_id',\n",
       " 'development_stage',\n",
       " 'anatomical_region',\n",
       " 'anatomical_region_merged',\n",
       " 'anatomical_region_ontology_term_id',\n",
       " 'brain_region_ontology_term_id',\n",
       " 'self_reported_sex',\n",
       " 'self_reported_sex_ontology_term_id',\n",
       " 'self_reported_ethnicity',\n",
       " 'self_reported_ethnicity_ontology_term_id',\n",
       " 'disease',\n",
       " 'disease_ontology_term_id',\n",
       " 'suspension_type',\n",
       " 'is_primary_data',\n",
       " 'atac_confidently_mapped_read_pairs',\n",
       " 'atac_fraction_of_genome_in_peaks',\n",
       " 'atac_fraction_of_high_quality_fragments_in_cells',\n",
       " 'atac_fraction_of_high_quality_fragments_overlapping_tss',\n",
       " 'atac_fraction_of_high_quality_fragments_overlapping_peaks',\n",
       " 'atac_fraction_of_transposition_events_in_peaks_in_cells',\n",
       " 'atac_mean_raw_read_pairs_per_cell',\n",
       " 'atac_median_high_quality_fragments_per_cell',\n",
       " 'atac_non_nuclear_read_pairs',\n",
       " 'atac_number_of_peaks',\n",
       " 'atac_percent_duplicates',\n",
       " 'atac_q30_bases_in_barcode',\n",
       " 'atac_q30_bases_in_read_1',\n",
       " 'atac_q30_bases_in_read_2',\n",
       " 'atac_q30_bases_in_sample_index_i1',\n",
       " 'atac_sequenced_read_pairs',\n",
       " 'atac_tss_enrichment_score',\n",
       " 'atac_unmapped_read_pairs',\n",
       " 'atac_valid_barcodes',\n",
       " 'gex_fraction_of_transcriptomic_reads_in_cells',\n",
       " 'gex_mean_raw_reads_per_cell',\n",
       " 'gex_median_umi_counts_per_cell',\n",
       " 'gex_median_genes_per_cell',\n",
       " 'gex_percent_duplicates',\n",
       " 'gex_q30_bases_in_umi',\n",
       " 'gex_q30_bases_in_barcode',\n",
       " 'gex_q30_bases_in_read_2',\n",
       " 'gex_q30_bases_in_sample_index_i1',\n",
       " 'gex_q30_bases_in_sample_index_i2',\n",
       " 'gex_reads_mapped_antisense_to_gene',\n",
       " 'gex_reads_mapped_confidently_to_exonic_regions',\n",
       " 'gex_reads_mapped_confidently_to_genome',\n",
       " 'gex_reads_mapped_confidently_to_intergenic_regions',\n",
       " 'gex_reads_mapped_confidently_to_intronic_regions',\n",
       " 'gex_reads_mapped_confidently_to_transcriptome',\n",
       " 'gex_reads_mapped_to_genome',\n",
       " 'gex_reads_with_tso',\n",
       " 'gex_sequenced_read_pairs',\n",
       " 'gex_total_genes_detected',\n",
       " 'gex_valid_umis',\n",
       " 'gex_valid_barcodes',\n",
       " 'total_counts',\n",
       " 'total_genes',\n",
       " 'doublet_score',\n",
       " 'organism_sci',\n",
       " 'accession_group',\n",
       " 'embedding_set',\n",
       " 'CL:ID_group',\n",
       " 'tokens_group',\n",
       " 'display_order_group',\n",
       " 'literature_support',\n",
       " 'literature_name_short',\n",
       " 'literature_name_long',\n",
       " 'curated_markers',\n",
       " 'color_hex_group',\n",
       " 'color_group',\n",
       " 'accession_subclass',\n",
       " 'CL:ID_subclass',\n",
       " 'tokens_subclass',\n",
       " 'display_order_subclass',\n",
       " 'color_hex_subclass',\n",
       " 'color_subclass',\n",
       " 'accession_class',\n",
       " 'CL:ID_class',\n",
       " 'tokens_class',\n",
       " 'display_order_class',\n",
       " 'color_hex_class',\n",
       " 'color_class',\n",
       " 'accession_neighborhood',\n",
       " 'CL:ID_neighborhood',\n",
       " 'tokens_neighborhood',\n",
       " 'display_order_neighborhood',\n",
       " 'color_hex_neighborhood',\n",
       " 'color_neighborhood',\n",
       " 'cell_barcode',\n",
       " 'barcoded_cell_sample_label',\n",
       " 'alignment_job_database']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(ann_data.obs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f15650c-417d-4c5c-9998-e9e275de87d3",
   "metadata": {},
   "source": [
    "Rename `cell_type_ontology_term` to `cell_type_ontology_term_id` to be compatible with the [CxG schema](https://github.com/chanzuckerberg/single-cell-curation/blob/main/schema/4.0.0/schema.md#cell_type_ontology_term_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "37e882d4-6015-4287-bf4c-f0e20f223b9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "ann_data.obs.rename(columns={'cell_type_ontology_term': 'cell_type_ontology_term_id'}, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e9c0c79-9e1d-425e-9f94-54f9170acb08",
   "metadata": {},
   "source": [
    "Add a new `cell_type` column to be compatible with the [CxG schema](https://github.com/chanzuckerberg/single-cell-curation/blob/main/schema/4.0.0/schema.md#cell_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8116b0a5-bca6-4fa2-815a-14ad394d4b8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import ssl\n",
    "from SPARQLWrapper import SPARQLWrapper, JSON\n",
    "from cachetools import TTLCache, cached\n",
    "\n",
    "cache = TTLCache(maxsize=256, ttl=3600)\n",
    "ssl._create_default_https_context = ssl._create_unverified_context\n",
    "\n",
    "@cached(cache)\n",
    "def get_ontology_label(term_curie: str) -> str:\n",
    "    term_id = term_curie.strip().replace(\":\", \"_\")\n",
    "    sparql = SPARQLWrapper(\"https://ubergraph.apps.renci.org/sparql\")\n",
    "    query = \"\"\"\n",
    "    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
    "    PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "    SELECT DISTINCT *\n",
    "    WHERE {{\n",
    "      obo:{term_id} rdfs:label ?label .\n",
    "    }}\n",
    "    LIMIT 1\n",
    "    \"\"\".format(term_id=term_id)\n",
    "    sparql.setQuery(query)\n",
    "    sparql.setReturnFormat(JSON)\n",
    "    results = sparql.query().convert()\n",
    "\n",
    "    bindings = results.get(\"results\", {}).get(\"bindings\", [])\n",
    "    if not bindings:\n",
    "        raise Exception(\"Ontology term not found: {}\".format(term_curie))\n",
    "\n",
    "    label = bindings[0].get(\"label\", {}).get(\"value\")\n",
    "    return label\n",
    "\n",
    "ann_data.obs['cell_type'] = ann_data.obs['cell_type_ontology_term_id'].apply(get_ontology_label)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e0250be-cbfa-43ac-a5b9-e99b28428537",
   "metadata": {},
   "source": [
    "Add `accession_cluster` column to the obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3eaf40b8-f045-4129-95a3-74ff6921a8a4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       accession_cluster  cluster_id     Neighborhood     Class Subclass  \\\n",
      "0  CS20250428_CLUST_0569   Macaque-7  Subpallium GABA  F M GABA   F GABA   \n",
      "1  CS20250428_CLUST_0570   Macaque-8  Subpallium GABA  F M GABA   F GABA   \n",
      "2  CS20250428_CLUST_0600  Macaque-18  Subpallium GABA  F M GABA   F GABA   \n",
      "3  CS20250428_CLUST_0612  Macaque-30  Subpallium GABA  F M GABA   F GABA   \n",
      "4  CS20250428_CLUST_0605  Macaque-23  Subpallium GABA  F M GABA   F GABA   \n",
      "\n",
      "                   Group  \n",
      "0  AMY-SLEA-BNST D1 GABA  \n",
      "1  AMY-SLEA-BNST D1 GABA  \n",
      "2     AMY-SLEA-BNST GABA  \n",
      "3     AMY-SLEA-BNST GABA  \n",
      "4     AMY-SLEA-BNST GABA  \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "cluster_id\n",
       "Macaque-7      CS20250428_CLUST_0569\n",
       "Macaque-8      CS20250428_CLUST_0570\n",
       "Macaque-18     CS20250428_CLUST_0600\n",
       "Macaque-30     CS20250428_CLUST_0612\n",
       "Macaque-23     CS20250428_CLUST_0605\n",
       "                       ...          \n",
       "Macaque-296    CS20250428_CLUST_1058\n",
       "Macaque-311    CS20250428_CLUST_1068\n",
       "Macaque-298    CS20250428_CLUST_1060\n",
       "Macaque-302    CS20250428_CLUST_1062\n",
       "Macaque-314    CS20250428_CLUST_1071\n",
       "Name: accession_cluster, Length: 388, dtype: object"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "url = \"https://docs.google.com/spreadsheets/d/1kXxOam_MddplC5UUikGhvdDX0rF2eslaGFni0KhjybQ/export?format=csv&gid=563572258\"\n",
    "annotation_table = pd.read_csv(url)\n",
    "print(annotation_table.head())\n",
    "mapping = annotation_table.set_index('cluster_id')['accession_cluster']\n",
    "mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "556e0414-cd09-46f1-a681-78c63d61e4b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "ann_data.obs['accession_cluster'] = ann_data.obs['cluster_id'].map(mapping)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f86f61b5-8bf7-4e61-ba85-a5a801007692",
   "metadata": {},
   "source": [
    "Save the updated AnnData file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "630aaba6-f20e-4084-af1b-d17b3d1d9568",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Neighborhood</th>\n",
       "      <th>Class</th>\n",
       "      <th>Subclass</th>\n",
       "      <th>Group</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>cell_type_ontology_term_id</th>\n",
       "      <th>load_id</th>\n",
       "      <th>donor_id</th>\n",
       "      <th>assay</th>\n",
       "      <th>assay_ontology_term_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>organism_ontology_term_id</th>\n",
       "      <th>development_stage</th>\n",
       "      <th>anatomical_region</th>\n",
       "      <th>anatomical_region_merged</th>\n",
       "      <th>anatomical_region_ontology_term_id</th>\n",
       "      <th>brain_region_ontology_term_id</th>\n",
       "      <th>self_reported_sex</th>\n",
       "      <th>self_reported_sex_ontology_term_id</th>\n",
       "      <th>self_reported_ethnicity</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>is_primary_data</th>\n",
       "      <th>atac_confidently_mapped_read_pairs</th>\n",
       "      <th>atac_fraction_of_genome_in_peaks</th>\n",
       "      <th>atac_fraction_of_high_quality_fragments_in_cells</th>\n",
       "      <th>atac_fraction_of_high_quality_fragments_overlapping_tss</th>\n",
       "      <th>atac_fraction_of_high_quality_fragments_overlapping_peaks</th>\n",
       "      <th>atac_fraction_of_transposition_events_in_peaks_in_cells</th>\n",
       "      <th>atac_mean_raw_read_pairs_per_cell</th>\n",
       "      <th>atac_median_high_quality_fragments_per_cell</th>\n",
       "      <th>atac_non_nuclear_read_pairs</th>\n",
       "      <th>atac_number_of_peaks</th>\n",
       "      <th>atac_percent_duplicates</th>\n",
       "      <th>atac_q30_bases_in_barcode</th>\n",
       "      <th>atac_q30_bases_in_read_1</th>\n",
       "      <th>atac_q30_bases_in_read_2</th>\n",
       "      <th>atac_q30_bases_in_sample_index_i1</th>\n",
       "      <th>atac_sequenced_read_pairs</th>\n",
       "      <th>atac_tss_enrichment_score</th>\n",
       "      <th>atac_unmapped_read_pairs</th>\n",
       "      <th>atac_valid_barcodes</th>\n",
       "      <th>gex_fraction_of_transcriptomic_reads_in_cells</th>\n",
       "      <th>gex_mean_raw_reads_per_cell</th>\n",
       "      <th>gex_median_umi_counts_per_cell</th>\n",
       "      <th>gex_median_genes_per_cell</th>\n",
       "      <th>gex_percent_duplicates</th>\n",
       "      <th>gex_q30_bases_in_umi</th>\n",
       "      <th>gex_q30_bases_in_barcode</th>\n",
       "      <th>gex_q30_bases_in_read_2</th>\n",
       "      <th>gex_q30_bases_in_sample_index_i1</th>\n",
       "      <th>gex_q30_bases_in_sample_index_i2</th>\n",
       "      <th>gex_reads_mapped_antisense_to_gene</th>\n",
       "      <th>gex_reads_mapped_confidently_to_exonic_regions</th>\n",
       "      <th>gex_reads_mapped_confidently_to_genome</th>\n",
       "      <th>gex_reads_mapped_confidently_to_intergenic_regions</th>\n",
       "      <th>gex_reads_mapped_confidently_to_intronic_regions</th>\n",
       "      <th>gex_reads_mapped_confidently_to_transcriptome</th>\n",
       "      <th>gex_reads_mapped_to_genome</th>\n",
       "      <th>gex_reads_with_tso</th>\n",
       "      <th>gex_sequenced_read_pairs</th>\n",
       "      <th>gex_total_genes_detected</th>\n",
       "      <th>gex_valid_umis</th>\n",
       "      <th>gex_valid_barcodes</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>total_genes</th>\n",
       "      <th>doublet_score</th>\n",
       "      <th>organism_sci</th>\n",
       "      <th>accession_group</th>\n",
       "      <th>embedding_set</th>\n",
       "      <th>CL:ID_group</th>\n",
       "      <th>tokens_group</th>\n",
       "      <th>display_order_group</th>\n",
       "      <th>literature_support</th>\n",
       "      <th>literature_name_short</th>\n",
       "      <th>literature_name_long</th>\n",
       "      <th>curated_markers</th>\n",
       "      <th>color_hex_group</th>\n",
       "      <th>color_group</th>\n",
       "      <th>accession_subclass</th>\n",
       "      <th>CL:ID_subclass</th>\n",
       "      <th>tokens_subclass</th>\n",
       "      <th>display_order_subclass</th>\n",
       "      <th>color_hex_subclass</th>\n",
       "      <th>color_subclass</th>\n",
       "      <th>accession_class</th>\n",
       "      <th>CL:ID_class</th>\n",
       "      <th>tokens_class</th>\n",
       "      <th>display_order_class</th>\n",
       "      <th>color_hex_class</th>\n",
       "      <th>color_class</th>\n",
       "      <th>accession_neighborhood</th>\n",
       "      <th>CL:ID_neighborhood</th>\n",
       "      <th>tokens_neighborhood</th>\n",
       "      <th>display_order_neighborhood</th>\n",
       "      <th>color_hex_neighborhood</th>\n",
       "      <th>color_neighborhood</th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>barcoded_cell_sample_label</th>\n",
       "      <th>alignment_job_database</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>accession_cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACAGCCAGTTGCGT-1061_B04</th>\n",
       "      <td>Subpallium GABA</td>\n",
       "      <td>CN LGE GABA</td>\n",
       "      <td>STR D1 MSN</td>\n",
       "      <td>STRv D1 MSN</td>\n",
       "      <td>Macaque-1</td>\n",
       "      <td>Macaque-1</td>\n",
       "      <td>CL:0000003</td>\n",
       "      <td>1061_B04</td>\n",
       "      <td>Q21.26.010</td>\n",
       "      <td>10x multiome</td>\n",
       "      <td>EFO_0030059</td>\n",
       "      <td>Macaque</td>\n",
       "      <td>NCBITaxon:9544</td>\n",
       "      <td>Adult</td>\n",
       "      <td>NAC</td>\n",
       "      <td>NAC</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Male</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Macaca Mulatta</td>\n",
       "      <td>normal</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>True</td>\n",
       "      <td>0.8805</td>\n",
       "      <td>0.028</td>\n",
       "      <td>0.7524</td>\n",
       "      <td>0.213</td>\n",
       "      <td>0.2391</td>\n",
       "      <td>0.2184</td>\n",
       "      <td>95529.2519</td>\n",
       "      <td>35407.0</td>\n",
       "      <td>0</td>\n",
       "      <td>95381</td>\n",
       "      <td>0.3599</td>\n",
       "      <td>0.9235</td>\n",
       "      <td>0.9632</td>\n",
       "      <td>0.9532</td>\n",
       "      <td>0.9523</td>\n",
       "      <td>525315356</td>\n",
       "      <td>4.6613</td>\n",
       "      <td>0.0107</td>\n",
       "      <td>0.9846</td>\n",
       "      <td>0.9301</td>\n",
       "      <td>42172.1289</td>\n",
       "      <td>10016.0</td>\n",
       "      <td>3875.0</td>\n",
       "      <td>0.3316</td>\n",
       "      <td>0.9387</td>\n",
       "      <td>0.9398</td>\n",
       "      <td>0.9272</td>\n",
       "      <td>0.9373</td>\n",
       "      <td>0.9069</td>\n",
       "      <td>0.341</td>\n",
       "      <td>0.1172</td>\n",
       "      <td>0.9076</td>\n",
       "      <td>0.1177</td>\n",
       "      <td>0.6727</td>\n",
       "      <td>0.4463</td>\n",
       "      <td>0.9485</td>\n",
       "      <td>0.1083</td>\n",
       "      <td>231904537</td>\n",
       "      <td>29345</td>\n",
       "      <td>0.9994</td>\n",
       "      <td>0.892</td>\n",
       "      <td>14195</td>\n",
       "      <td>4639</td>\n",
       "      <td>0.018182</td>\n",
       "      <td>Macaca Mulatta</td>\n",
       "      <td>CS20250428_GROUP_0056</td>\n",
       "      <td>LGE</td>\n",
       "      <td>CL:4030051</td>\n",
       "      <td>STRv|D1 MSN</td>\n",
       "      <td>51</td>\n",
       "      <td>He et al. 2021</td>\n",
       "      <td>D1-Shell/OT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ARHGAP6</td>\n",
       "      <td>#339933</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CS20250428_SUBCL_0028</td>\n",
       "      <td>CL:4023026</td>\n",
       "      <td>STR|D1 MSN</td>\n",
       "      <td>31</td>\n",
       "      <td>#1655f2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CS20250428_CLASS_0003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CN|LGE|GABA</td>\n",
       "      <td>10</td>\n",
       "      <td>#d0b83c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CS20250428_NEIGH_0002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Subpallium|GABA</td>\n",
       "      <td>3</td>\n",
       "      <td>#19613b</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AAACAGCCAGTTGCGT</td>\n",
       "      <td>1061_B04</td>\n",
       "      <td>AIBS</td>\n",
       "      <td>obsolete native cell</td>\n",
       "      <td>CS20250428_CLUST_0991</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATGCAATTGACT-1061_B04</th>\n",
       "      <td>Subpallium GABA</td>\n",
       "      <td>F M GABA</td>\n",
       "      <td>F GABA</td>\n",
       "      <td>AMY-SLEA-BNST GABA</td>\n",
       "      <td>Macaque-18</td>\n",
       "      <td>Macaque-18</td>\n",
       "      <td>CL:0000003</td>\n",
       "      <td>1061_B04</td>\n",
       "      <td>Q21.26.010</td>\n",
       "      <td>10x multiome</td>\n",
       "      <td>EFO_0030059</td>\n",
       "      <td>Macaque</td>\n",
       "      <td>NCBITaxon:9544</td>\n",
       "      <td>Adult</td>\n",
       "      <td>NAC</td>\n",
       "      <td>NAC</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Male</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Macaca Mulatta</td>\n",
       "      <td>normal</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>True</td>\n",
       "      <td>0.8805</td>\n",
       "      <td>0.028</td>\n",
       "      <td>0.7524</td>\n",
       "      <td>0.213</td>\n",
       "      <td>0.2391</td>\n",
       "      <td>0.2184</td>\n",
       "      <td>95529.2519</td>\n",
       "      <td>35407.0</td>\n",
       "      <td>0</td>\n",
       "      <td>95381</td>\n",
       "      <td>0.3599</td>\n",
       "      <td>0.9235</td>\n",
       "      <td>0.9632</td>\n",
       "      <td>0.9532</td>\n",
       "      <td>0.9523</td>\n",
       "      <td>525315356</td>\n",
       "      <td>4.6613</td>\n",
       "      <td>0.0107</td>\n",
       "      <td>0.9846</td>\n",
       "      <td>0.9301</td>\n",
       "      <td>42172.1289</td>\n",
       "      <td>10016.0</td>\n",
       "      <td>3875.0</td>\n",
       "      <td>0.3316</td>\n",
       "      <td>0.9387</td>\n",
       "      <td>0.9398</td>\n",
       "      <td>0.9272</td>\n",
       "      <td>0.9373</td>\n",
       "      <td>0.9069</td>\n",
       "      <td>0.341</td>\n",
       "      <td>0.1172</td>\n",
       "      <td>0.9076</td>\n",
       "      <td>0.1177</td>\n",
       "      <td>0.6727</td>\n",
       "      <td>0.4463</td>\n",
       "      <td>0.9485</td>\n",
       "      <td>0.1083</td>\n",
       "      <td>231904537</td>\n",
       "      <td>29345</td>\n",
       "      <td>0.9994</td>\n",
       "      <td>0.892</td>\n",
       "      <td>4441</td>\n",
       "      <td>2507</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Macaca Mulatta</td>\n",
       "      <td>CS20250428_GROUP_0066</td>\n",
       "      <td>Interneuron</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AMY|SLEA|BNST|GABA</td>\n",
       "      <td>48</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>#b69e14</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CS20250428_SUBCL_0037</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F|GABA</td>\n",
       "      <td>29</td>\n",
       "      <td>#7ba2df</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CS20250428_CLASS_0006</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F|M|GABA</td>\n",
       "      <td>9</td>\n",
       "      <td>#cd0f13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CS20250428_NEIGH_0002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Subpallium|GABA</td>\n",
       "      <td>3</td>\n",
       "      <td>#19613b</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AAACATGCAATTGACT</td>\n",
       "      <td>1061_B04</td>\n",
       "      <td>AIBS</td>\n",
       "      <td>obsolete native cell</td>\n",
       "      <td>CS20250428_CLUST_0600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATGCACGTTACA-1061_B04</th>\n",
       "      <td>Subpallium GABA</td>\n",
       "      <td>CN LGE GABA</td>\n",
       "      <td>STR Hybrid MSN</td>\n",
       "      <td>STRv D1 NUDAP MSN</td>\n",
       "      <td>Macaque-35</td>\n",
       "      <td>Macaque-35</td>\n",
       "      <td>CL:0000003</td>\n",
       "      <td>1061_B04</td>\n",
       "      <td>Q21.26.010</td>\n",
       "      <td>10x multiome</td>\n",
       "      <td>EFO_0030059</td>\n",
       "      <td>Macaque</td>\n",
       "      <td>NCBITaxon:9544</td>\n",
       "      <td>Adult</td>\n",
       "      <td>NAC</td>\n",
       "      <td>NAC</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Male</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Macaca Mulatta</td>\n",
       "      <td>normal</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>True</td>\n",
       "      <td>0.8805</td>\n",
       "      <td>0.028</td>\n",
       "      <td>0.7524</td>\n",
       "      <td>0.213</td>\n",
       "      <td>0.2391</td>\n",
       "      <td>0.2184</td>\n",
       "      <td>95529.2519</td>\n",
       "      <td>35407.0</td>\n",
       "      <td>0</td>\n",
       "      <td>95381</td>\n",
       "      <td>0.3599</td>\n",
       "      <td>0.9235</td>\n",
       "      <td>0.9632</td>\n",
       "      <td>0.9532</td>\n",
       "      <td>0.9523</td>\n",
       "      <td>525315356</td>\n",
       "      <td>4.6613</td>\n",
       "      <td>0.0107</td>\n",
       "      <td>0.9846</td>\n",
       "      <td>0.9301</td>\n",
       "      <td>42172.1289</td>\n",
       "      <td>10016.0</td>\n",
       "      <td>3875.0</td>\n",
       "      <td>0.3316</td>\n",
       "      <td>0.9387</td>\n",
       "      <td>0.9398</td>\n",
       "      <td>0.9272</td>\n",
       "      <td>0.9373</td>\n",
       "      <td>0.9069</td>\n",
       "      <td>0.341</td>\n",
       "      <td>0.1172</td>\n",
       "      <td>0.9076</td>\n",
       "      <td>0.1177</td>\n",
       "      <td>0.6727</td>\n",
       "      <td>0.4463</td>\n",
       "      <td>0.9485</td>\n",
       "      <td>0.1083</td>\n",
       "      <td>231904537</td>\n",
       "      <td>29345</td>\n",
       "      <td>0.9994</td>\n",
       "      <td>0.892</td>\n",
       "      <td>10473</td>\n",
       "      <td>3808</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Macaca Mulatta</td>\n",
       "      <td>CS20250428_GROUP_0057</td>\n",
       "      <td>LGE</td>\n",
       "      <td>CL:4030054</td>\n",
       "      <td>STRv|D1 NUDAP MSN</td>\n",
       "      <td>57</td>\n",
       "      <td>He et al. 2021</td>\n",
       "      <td>D1-NUDAP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GRIA4</td>\n",
       "      <td>#e377c2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CS20250428_SUBCL_0030</td>\n",
       "      <td>CL:4030050</td>\n",
       "      <td>STR|Hybrid MSN</td>\n",
       "      <td>33</td>\n",
       "      <td>#79bdf4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CS20250428_CLASS_0003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CN|LGE|GABA</td>\n",
       "      <td>10</td>\n",
       "      <td>#d0b83c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CS20250428_NEIGH_0002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Subpallium|GABA</td>\n",
       "      <td>3</td>\n",
       "      <td>#19613b</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AAACATGCACGTTACA</td>\n",
       "      <td>1061_B04</td>\n",
       "      <td>AIBS</td>\n",
       "      <td>obsolete native cell</td>\n",
       "      <td>CS20250428_CLUST_1019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATGCAGGTTACC-1061_B04</th>\n",
       "      <td>Subpallium GABA</td>\n",
       "      <td>CN LGE GABA</td>\n",
       "      <td>STR Hybrid MSN</td>\n",
       "      <td>STRv D1 NUDAP MSN</td>\n",
       "      <td>Macaque-31</td>\n",
       "      <td>Macaque-31</td>\n",
       "      <td>CL:0000003</td>\n",
       "      <td>1061_B04</td>\n",
       "      <td>Q21.26.010</td>\n",
       "      <td>10x multiome</td>\n",
       "      <td>EFO_0030059</td>\n",
       "      <td>Macaque</td>\n",
       "      <td>NCBITaxon:9544</td>\n",
       "      <td>Adult</td>\n",
       "      <td>NAC</td>\n",
       "      <td>NAC</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Male</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Macaca Mulatta</td>\n",
       "      <td>normal</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>True</td>\n",
       "      <td>0.8805</td>\n",
       "      <td>0.028</td>\n",
       "      <td>0.7524</td>\n",
       "      <td>0.213</td>\n",
       "      <td>0.2391</td>\n",
       "      <td>0.2184</td>\n",
       "      <td>95529.2519</td>\n",
       "      <td>35407.0</td>\n",
       "      <td>0</td>\n",
       "      <td>95381</td>\n",
       "      <td>0.3599</td>\n",
       "      <td>0.9235</td>\n",
       "      <td>0.9632</td>\n",
       "      <td>0.9532</td>\n",
       "      <td>0.9523</td>\n",
       "      <td>525315356</td>\n",
       "      <td>4.6613</td>\n",
       "      <td>0.0107</td>\n",
       "      <td>0.9846</td>\n",
       "      <td>0.9301</td>\n",
       "      <td>42172.1289</td>\n",
       "      <td>10016.0</td>\n",
       "      <td>3875.0</td>\n",
       "      <td>0.3316</td>\n",
       "      <td>0.9387</td>\n",
       "      <td>0.9398</td>\n",
       "      <td>0.9272</td>\n",
       "      <td>0.9373</td>\n",
       "      <td>0.9069</td>\n",
       "      <td>0.341</td>\n",
       "      <td>0.1172</td>\n",
       "      <td>0.9076</td>\n",
       "      <td>0.1177</td>\n",
       "      <td>0.6727</td>\n",
       "      <td>0.4463</td>\n",
       "      <td>0.9485</td>\n",
       "      <td>0.1083</td>\n",
       "      <td>231904537</td>\n",
       "      <td>29345</td>\n",
       "      <td>0.9994</td>\n",
       "      <td>0.892</td>\n",
       "      <td>11817</td>\n",
       "      <td>4080</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Macaca Mulatta</td>\n",
       "      <td>CS20250428_GROUP_0057</td>\n",
       "      <td>LGE</td>\n",
       "      <td>CL:4030054</td>\n",
       "      <td>STRv|D1 NUDAP MSN</td>\n",
       "      <td>57</td>\n",
       "      <td>He et al. 2021</td>\n",
       "      <td>D1-NUDAP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GRIA4</td>\n",
       "      <td>#e377c2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CS20250428_SUBCL_0030</td>\n",
       "      <td>CL:4030050</td>\n",
       "      <td>STR|Hybrid MSN</td>\n",
       "      <td>33</td>\n",
       "      <td>#79bdf4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CS20250428_CLASS_0003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CN|LGE|GABA</td>\n",
       "      <td>10</td>\n",
       "      <td>#d0b83c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CS20250428_NEIGH_0002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Subpallium|GABA</td>\n",
       "      <td>3</td>\n",
       "      <td>#19613b</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AAACATGCAGGTTACC</td>\n",
       "      <td>1061_B04</td>\n",
       "      <td>AIBS</td>\n",
       "      <td>obsolete native cell</td>\n",
       "      <td>CS20250428_CLUST_1015</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              Neighborhood        Class        Subclass  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04  Subpallium GABA  CN LGE GABA      STR D1 MSN   \n",
       "AAACATGCAATTGACT-1061_B04  Subpallium GABA     F M GABA          F GABA   \n",
       "AAACATGCACGTTACA-1061_B04  Subpallium GABA  CN LGE GABA  STR Hybrid MSN   \n",
       "AAACATGCAGGTTACC-1061_B04  Subpallium GABA  CN LGE GABA  STR Hybrid MSN   \n",
       "\n",
       "                                        Group     Cluster  cluster_id  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04         STRv D1 MSN   Macaque-1   Macaque-1   \n",
       "AAACATGCAATTGACT-1061_B04  AMY-SLEA-BNST GABA  Macaque-18  Macaque-18   \n",
       "AAACATGCACGTTACA-1061_B04   STRv D1 NUDAP MSN  Macaque-35  Macaque-35   \n",
       "AAACATGCAGGTTACC-1061_B04   STRv D1 NUDAP MSN  Macaque-31  Macaque-31   \n",
       "\n",
       "                          cell_type_ontology_term_id   load_id    donor_id  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                 CL:0000003  1061_B04  Q21.26.010   \n",
       "AAACATGCAATTGACT-1061_B04                 CL:0000003  1061_B04  Q21.26.010   \n",
       "AAACATGCACGTTACA-1061_B04                 CL:0000003  1061_B04  Q21.26.010   \n",
       "AAACATGCAGGTTACC-1061_B04                 CL:0000003  1061_B04  Q21.26.010   \n",
       "\n",
       "                                  assay assay_ontology_term_id organism  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04  10x multiome            EFO_0030059  Macaque   \n",
       "AAACATGCAATTGACT-1061_B04  10x multiome            EFO_0030059  Macaque   \n",
       "AAACATGCACGTTACA-1061_B04  10x multiome            EFO_0030059  Macaque   \n",
       "AAACATGCAGGTTACC-1061_B04  10x multiome            EFO_0030059  Macaque   \n",
       "\n",
       "                          organism_ontology_term_id development_stage  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04            NCBITaxon:9544             Adult   \n",
       "AAACATGCAATTGACT-1061_B04            NCBITaxon:9544             Adult   \n",
       "AAACATGCACGTTACA-1061_B04            NCBITaxon:9544             Adult   \n",
       "AAACATGCAGGTTACC-1061_B04            NCBITaxon:9544             Adult   \n",
       "\n",
       "                          anatomical_region anatomical_region_merged  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04               NAC                      NAC   \n",
       "AAACATGCAATTGACT-1061_B04               NAC                      NAC   \n",
       "AAACATGCACGTTACA-1061_B04               NAC                      NAC   \n",
       "AAACATGCAGGTTACC-1061_B04               NAC                      NAC   \n",
       "\n",
       "                          anatomical_region_ontology_term_id  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                            Unknown   \n",
       "AAACATGCAATTGACT-1061_B04                            Unknown   \n",
       "AAACATGCACGTTACA-1061_B04                            Unknown   \n",
       "AAACATGCAGGTTACC-1061_B04                            Unknown   \n",
       "\n",
       "                          brain_region_ontology_term_id self_reported_sex  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                       Unknown              Male   \n",
       "AAACATGCAATTGACT-1061_B04                       Unknown              Male   \n",
       "AAACATGCACGTTACA-1061_B04                       Unknown              Male   \n",
       "AAACATGCAGGTTACC-1061_B04                       Unknown              Male   \n",
       "\n",
       "                          self_reported_sex_ontology_term_id  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                       PATO:0000384   \n",
       "AAACATGCAATTGACT-1061_B04                       PATO:0000384   \n",
       "AAACATGCACGTTACA-1061_B04                       PATO:0000384   \n",
       "AAACATGCAGGTTACC-1061_B04                       PATO:0000384   \n",
       "\n",
       "                          self_reported_ethnicity  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                 unknown   \n",
       "AAACATGCAATTGACT-1061_B04                 unknown   \n",
       "AAACATGCACGTTACA-1061_B04                 unknown   \n",
       "AAACATGCAGGTTACC-1061_B04                 unknown   \n",
       "\n",
       "                          self_reported_ethnicity_ontology_term_id disease  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                           Macaca Mulatta  normal   \n",
       "AAACATGCAATTGACT-1061_B04                           Macaca Mulatta  normal   \n",
       "AAACATGCACGTTACA-1061_B04                           Macaca Mulatta  normal   \n",
       "AAACATGCAGGTTACC-1061_B04                           Macaca Mulatta  normal   \n",
       "\n",
       "                          disease_ontology_term_id suspension_type  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04             PATO:0000461         nucleus   \n",
       "AAACATGCAATTGACT-1061_B04             PATO:0000461         nucleus   \n",
       "AAACATGCACGTTACA-1061_B04             PATO:0000461         nucleus   \n",
       "AAACATGCAGGTTACC-1061_B04             PATO:0000461         nucleus   \n",
       "\n",
       "                          is_primary_data  atac_confidently_mapped_read_pairs  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04            True                              0.8805   \n",
       "AAACATGCAATTGACT-1061_B04            True                              0.8805   \n",
       "AAACATGCACGTTACA-1061_B04            True                              0.8805   \n",
       "AAACATGCAGGTTACC-1061_B04            True                              0.8805   \n",
       "\n",
       "                           atac_fraction_of_genome_in_peaks  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                             0.028   \n",
       "AAACATGCAATTGACT-1061_B04                             0.028   \n",
       "AAACATGCACGTTACA-1061_B04                             0.028   \n",
       "AAACATGCAGGTTACC-1061_B04                             0.028   \n",
       "\n",
       "                           atac_fraction_of_high_quality_fragments_in_cells  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                                            0.7524   \n",
       "AAACATGCAATTGACT-1061_B04                                            0.7524   \n",
       "AAACATGCACGTTACA-1061_B04                                            0.7524   \n",
       "AAACATGCAGGTTACC-1061_B04                                            0.7524   \n",
       "\n",
       "                           atac_fraction_of_high_quality_fragments_overlapping_tss  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                                              0.213         \n",
       "AAACATGCAATTGACT-1061_B04                                              0.213         \n",
       "AAACATGCACGTTACA-1061_B04                                              0.213         \n",
       "AAACATGCAGGTTACC-1061_B04                                              0.213         \n",
       "\n",
       "                           atac_fraction_of_high_quality_fragments_overlapping_peaks  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                                             0.2391           \n",
       "AAACATGCAATTGACT-1061_B04                                             0.2391           \n",
       "AAACATGCACGTTACA-1061_B04                                             0.2391           \n",
       "AAACATGCAGGTTACC-1061_B04                                             0.2391           \n",
       "\n",
       "                           atac_fraction_of_transposition_events_in_peaks_in_cells  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                                             0.2184         \n",
       "AAACATGCAATTGACT-1061_B04                                             0.2184         \n",
       "AAACATGCACGTTACA-1061_B04                                             0.2184         \n",
       "AAACATGCAGGTTACC-1061_B04                                             0.2184         \n",
       "\n",
       "                           atac_mean_raw_read_pairs_per_cell  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                         95529.2519   \n",
       "AAACATGCAATTGACT-1061_B04                         95529.2519   \n",
       "AAACATGCACGTTACA-1061_B04                         95529.2519   \n",
       "AAACATGCAGGTTACC-1061_B04                         95529.2519   \n",
       "\n",
       "                           atac_median_high_quality_fragments_per_cell  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                                      35407.0   \n",
       "AAACATGCAATTGACT-1061_B04                                      35407.0   \n",
       "AAACATGCACGTTACA-1061_B04                                      35407.0   \n",
       "AAACATGCAGGTTACC-1061_B04                                      35407.0   \n",
       "\n",
       "                           atac_non_nuclear_read_pairs  atac_number_of_peaks  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                            0                 95381   \n",
       "AAACATGCAATTGACT-1061_B04                            0                 95381   \n",
       "AAACATGCACGTTACA-1061_B04                            0                 95381   \n",
       "AAACATGCAGGTTACC-1061_B04                            0                 95381   \n",
       "\n",
       "                           atac_percent_duplicates  atac_q30_bases_in_barcode  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                   0.3599                     0.9235   \n",
       "AAACATGCAATTGACT-1061_B04                   0.3599                     0.9235   \n",
       "AAACATGCACGTTACA-1061_B04                   0.3599                     0.9235   \n",
       "AAACATGCAGGTTACC-1061_B04                   0.3599                     0.9235   \n",
       "\n",
       "                           atac_q30_bases_in_read_1  atac_q30_bases_in_read_2  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                    0.9632                    0.9532   \n",
       "AAACATGCAATTGACT-1061_B04                    0.9632                    0.9532   \n",
       "AAACATGCACGTTACA-1061_B04                    0.9632                    0.9532   \n",
       "AAACATGCAGGTTACC-1061_B04                    0.9632                    0.9532   \n",
       "\n",
       "                           atac_q30_bases_in_sample_index_i1  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                             0.9523   \n",
       "AAACATGCAATTGACT-1061_B04                             0.9523   \n",
       "AAACATGCACGTTACA-1061_B04                             0.9523   \n",
       "AAACATGCAGGTTACC-1061_B04                             0.9523   \n",
       "\n",
       "                           atac_sequenced_read_pairs  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                  525315356   \n",
       "AAACATGCAATTGACT-1061_B04                  525315356   \n",
       "AAACATGCACGTTACA-1061_B04                  525315356   \n",
       "AAACATGCAGGTTACC-1061_B04                  525315356   \n",
       "\n",
       "                           atac_tss_enrichment_score  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                     4.6613   \n",
       "AAACATGCAATTGACT-1061_B04                     4.6613   \n",
       "AAACATGCACGTTACA-1061_B04                     4.6613   \n",
       "AAACATGCAGGTTACC-1061_B04                     4.6613   \n",
       "\n",
       "                           atac_unmapped_read_pairs  atac_valid_barcodes  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                    0.0107               0.9846   \n",
       "AAACATGCAATTGACT-1061_B04                    0.0107               0.9846   \n",
       "AAACATGCACGTTACA-1061_B04                    0.0107               0.9846   \n",
       "AAACATGCAGGTTACC-1061_B04                    0.0107               0.9846   \n",
       "\n",
       "                           gex_fraction_of_transcriptomic_reads_in_cells  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                                         0.9301   \n",
       "AAACATGCAATTGACT-1061_B04                                         0.9301   \n",
       "AAACATGCACGTTACA-1061_B04                                         0.9301   \n",
       "AAACATGCAGGTTACC-1061_B04                                         0.9301   \n",
       "\n",
       "                           gex_mean_raw_reads_per_cell  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                   42172.1289   \n",
       "AAACATGCAATTGACT-1061_B04                   42172.1289   \n",
       "AAACATGCACGTTACA-1061_B04                   42172.1289   \n",
       "AAACATGCAGGTTACC-1061_B04                   42172.1289   \n",
       "\n",
       "                           gex_median_umi_counts_per_cell  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                         10016.0   \n",
       "AAACATGCAATTGACT-1061_B04                         10016.0   \n",
       "AAACATGCACGTTACA-1061_B04                         10016.0   \n",
       "AAACATGCAGGTTACC-1061_B04                         10016.0   \n",
       "\n",
       "                           gex_median_genes_per_cell  gex_percent_duplicates  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                     3875.0                  0.3316   \n",
       "AAACATGCAATTGACT-1061_B04                     3875.0                  0.3316   \n",
       "AAACATGCACGTTACA-1061_B04                     3875.0                  0.3316   \n",
       "AAACATGCAGGTTACC-1061_B04                     3875.0                  0.3316   \n",
       "\n",
       "                           gex_q30_bases_in_umi  gex_q30_bases_in_barcode  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                0.9387                    0.9398   \n",
       "AAACATGCAATTGACT-1061_B04                0.9387                    0.9398   \n",
       "AAACATGCACGTTACA-1061_B04                0.9387                    0.9398   \n",
       "AAACATGCAGGTTACC-1061_B04                0.9387                    0.9398   \n",
       "\n",
       "                           gex_q30_bases_in_read_2  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                   0.9272   \n",
       "AAACATGCAATTGACT-1061_B04                   0.9272   \n",
       "AAACATGCACGTTACA-1061_B04                   0.9272   \n",
       "AAACATGCAGGTTACC-1061_B04                   0.9272   \n",
       "\n",
       "                           gex_q30_bases_in_sample_index_i1  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                            0.9373   \n",
       "AAACATGCAATTGACT-1061_B04                            0.9373   \n",
       "AAACATGCACGTTACA-1061_B04                            0.9373   \n",
       "AAACATGCAGGTTACC-1061_B04                            0.9373   \n",
       "\n",
       "                           gex_q30_bases_in_sample_index_i2  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                            0.9069   \n",
       "AAACATGCAATTGACT-1061_B04                            0.9069   \n",
       "AAACATGCACGTTACA-1061_B04                            0.9069   \n",
       "AAACATGCAGGTTACC-1061_B04                            0.9069   \n",
       "\n",
       "                           gex_reads_mapped_antisense_to_gene  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                               0.341   \n",
       "AAACATGCAATTGACT-1061_B04                               0.341   \n",
       "AAACATGCACGTTACA-1061_B04                               0.341   \n",
       "AAACATGCAGGTTACC-1061_B04                               0.341   \n",
       "\n",
       "                           gex_reads_mapped_confidently_to_exonic_regions  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                                          0.1172   \n",
       "AAACATGCAATTGACT-1061_B04                                          0.1172   \n",
       "AAACATGCACGTTACA-1061_B04                                          0.1172   \n",
       "AAACATGCAGGTTACC-1061_B04                                          0.1172   \n",
       "\n",
       "                           gex_reads_mapped_confidently_to_genome  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                                  0.9076   \n",
       "AAACATGCAATTGACT-1061_B04                                  0.9076   \n",
       "AAACATGCACGTTACA-1061_B04                                  0.9076   \n",
       "AAACATGCAGGTTACC-1061_B04                                  0.9076   \n",
       "\n",
       "                           gex_reads_mapped_confidently_to_intergenic_regions  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                                             0.1177    \n",
       "AAACATGCAATTGACT-1061_B04                                             0.1177    \n",
       "AAACATGCACGTTACA-1061_B04                                             0.1177    \n",
       "AAACATGCAGGTTACC-1061_B04                                             0.1177    \n",
       "\n",
       "                           gex_reads_mapped_confidently_to_intronic_regions  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                                            0.6727   \n",
       "AAACATGCAATTGACT-1061_B04                                            0.6727   \n",
       "AAACATGCACGTTACA-1061_B04                                            0.6727   \n",
       "AAACATGCAGGTTACC-1061_B04                                            0.6727   \n",
       "\n",
       "                           gex_reads_mapped_confidently_to_transcriptome  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                                         0.4463   \n",
       "AAACATGCAATTGACT-1061_B04                                         0.4463   \n",
       "AAACATGCACGTTACA-1061_B04                                         0.4463   \n",
       "AAACATGCAGGTTACC-1061_B04                                         0.4463   \n",
       "\n",
       "                           gex_reads_mapped_to_genome  gex_reads_with_tso  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                      0.9485              0.1083   \n",
       "AAACATGCAATTGACT-1061_B04                      0.9485              0.1083   \n",
       "AAACATGCACGTTACA-1061_B04                      0.9485              0.1083   \n",
       "AAACATGCAGGTTACC-1061_B04                      0.9485              0.1083   \n",
       "\n",
       "                           gex_sequenced_read_pairs  gex_total_genes_detected  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                 231904537                     29345   \n",
       "AAACATGCAATTGACT-1061_B04                 231904537                     29345   \n",
       "AAACATGCACGTTACA-1061_B04                 231904537                     29345   \n",
       "AAACATGCAGGTTACC-1061_B04                 231904537                     29345   \n",
       "\n",
       "                           gex_valid_umis  gex_valid_barcodes  total_counts  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04          0.9994               0.892         14195   \n",
       "AAACATGCAATTGACT-1061_B04          0.9994               0.892          4441   \n",
       "AAACATGCACGTTACA-1061_B04          0.9994               0.892         10473   \n",
       "AAACATGCAGGTTACC-1061_B04          0.9994               0.892         11817   \n",
       "\n",
       "                           total_genes  doublet_score    organism_sci  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04         4639       0.018182  Macaca Mulatta   \n",
       "AAACATGCAATTGACT-1061_B04         2507       0.000000  Macaca Mulatta   \n",
       "AAACATGCACGTTACA-1061_B04         3808       0.000000  Macaca Mulatta   \n",
       "AAACATGCAGGTTACC-1061_B04         4080       0.000000  Macaca Mulatta   \n",
       "\n",
       "                                 accession_group embedding_set CL:ID_group  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04  CS20250428_GROUP_0056           LGE  CL:4030051   \n",
       "AAACATGCAATTGACT-1061_B04  CS20250428_GROUP_0066   Interneuron         NaN   \n",
       "AAACATGCACGTTACA-1061_B04  CS20250428_GROUP_0057           LGE  CL:4030054   \n",
       "AAACATGCAGGTTACC-1061_B04  CS20250428_GROUP_0057           LGE  CL:4030054   \n",
       "\n",
       "                                 tokens_group  display_order_group  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04         STRv|D1 MSN                   51   \n",
       "AAACATGCAATTGACT-1061_B04  AMY|SLEA|BNST|GABA                   48   \n",
       "AAACATGCACGTTACA-1061_B04   STRv|D1 NUDAP MSN                   57   \n",
       "AAACATGCAGGTTACC-1061_B04   STRv|D1 NUDAP MSN                   57   \n",
       "\n",
       "                          literature_support literature_name_short  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04     He et al. 2021           D1-Shell/OT   \n",
       "AAACATGCAATTGACT-1061_B04                NaN                   NaN   \n",
       "AAACATGCACGTTACA-1061_B04     He et al. 2021              D1-NUDAP   \n",
       "AAACATGCAGGTTACC-1061_B04     He et al. 2021              D1-NUDAP   \n",
       "\n",
       "                          literature_name_long curated_markers  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                  NaN         ARHGAP6   \n",
       "AAACATGCAATTGACT-1061_B04                  NaN             NaN   \n",
       "AAACATGCACGTTACA-1061_B04                  NaN           GRIA4   \n",
       "AAACATGCAGGTTACC-1061_B04                  NaN           GRIA4   \n",
       "\n",
       "                          color_hex_group  color_group     accession_subclass  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04         #339933          NaN  CS20250428_SUBCL_0028   \n",
       "AAACATGCAATTGACT-1061_B04         #b69e14          NaN  CS20250428_SUBCL_0037   \n",
       "AAACATGCACGTTACA-1061_B04         #e377c2          NaN  CS20250428_SUBCL_0030   \n",
       "AAACATGCAGGTTACC-1061_B04         #e377c2          NaN  CS20250428_SUBCL_0030   \n",
       "\n",
       "                          CL:ID_subclass tokens_subclass  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04     CL:4023026      STR|D1 MSN   \n",
       "AAACATGCAATTGACT-1061_B04            NaN          F|GABA   \n",
       "AAACATGCACGTTACA-1061_B04     CL:4030050  STR|Hybrid MSN   \n",
       "AAACATGCAGGTTACC-1061_B04     CL:4030050  STR|Hybrid MSN   \n",
       "\n",
       "                           display_order_subclass color_hex_subclass  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                      31            #1655f2   \n",
       "AAACATGCAATTGACT-1061_B04                      29            #7ba2df   \n",
       "AAACATGCACGTTACA-1061_B04                      33            #79bdf4   \n",
       "AAACATGCAGGTTACC-1061_B04                      33            #79bdf4   \n",
       "\n",
       "                           color_subclass        accession_class CL:ID_class  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04             NaN  CS20250428_CLASS_0003         NaN   \n",
       "AAACATGCAATTGACT-1061_B04             NaN  CS20250428_CLASS_0006         NaN   \n",
       "AAACATGCACGTTACA-1061_B04             NaN  CS20250428_CLASS_0003         NaN   \n",
       "AAACATGCAGGTTACC-1061_B04             NaN  CS20250428_CLASS_0003         NaN   \n",
       "\n",
       "                          tokens_class  display_order_class color_hex_class  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04  CN|LGE|GABA                   10         #d0b83c   \n",
       "AAACATGCAATTGACT-1061_B04     F|M|GABA                    9         #cd0f13   \n",
       "AAACATGCACGTTACA-1061_B04  CN|LGE|GABA                   10         #d0b83c   \n",
       "AAACATGCAGGTTACC-1061_B04  CN|LGE|GABA                   10         #d0b83c   \n",
       "\n",
       "                           color_class accession_neighborhood  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04          NaN  CS20250428_NEIGH_0002   \n",
       "AAACATGCAATTGACT-1061_B04          NaN  CS20250428_NEIGH_0002   \n",
       "AAACATGCACGTTACA-1061_B04          NaN  CS20250428_NEIGH_0002   \n",
       "AAACATGCAGGTTACC-1061_B04          NaN  CS20250428_NEIGH_0002   \n",
       "\n",
       "                          CL:ID_neighborhood tokens_neighborhood  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                NaN     Subpallium|GABA   \n",
       "AAACATGCAATTGACT-1061_B04                NaN     Subpallium|GABA   \n",
       "AAACATGCACGTTACA-1061_B04                NaN     Subpallium|GABA   \n",
       "AAACATGCAGGTTACC-1061_B04                NaN     Subpallium|GABA   \n",
       "\n",
       "                           display_order_neighborhood color_hex_neighborhood  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                           3                #19613b   \n",
       "AAACATGCAATTGACT-1061_B04                           3                #19613b   \n",
       "AAACATGCACGTTACA-1061_B04                           3                #19613b   \n",
       "AAACATGCAGGTTACC-1061_B04                           3                #19613b   \n",
       "\n",
       "                           color_neighborhood      cell_barcode  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                 NaN  AAACAGCCAGTTGCGT   \n",
       "AAACATGCAATTGACT-1061_B04                 NaN  AAACATGCAATTGACT   \n",
       "AAACATGCACGTTACA-1061_B04                 NaN  AAACATGCACGTTACA   \n",
       "AAACATGCAGGTTACC-1061_B04                 NaN  AAACATGCAGGTTACC   \n",
       "\n",
       "                          barcoded_cell_sample_label alignment_job_database  \\\n",
       "AAACAGCCAGTTGCGT-1061_B04                   1061_B04                   AIBS   \n",
       "AAACATGCAATTGACT-1061_B04                   1061_B04                   AIBS   \n",
       "AAACATGCACGTTACA-1061_B04                   1061_B04                   AIBS   \n",
       "AAACATGCAGGTTACC-1061_B04                   1061_B04                   AIBS   \n",
       "\n",
       "                                      cell_type      accession_cluster  \n",
       "AAACAGCCAGTTGCGT-1061_B04  obsolete native cell  CS20250428_CLUST_0991  \n",
       "AAACATGCAATTGACT-1061_B04  obsolete native cell  CS20250428_CLUST_0600  \n",
       "AAACATGCACGTTACA-1061_B04  obsolete native cell  CS20250428_CLUST_1019  \n",
       "AAACATGCAGGTTACC-1061_B04  obsolete native cell  CS20250428_CLUST_1015  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ann_data.obs.head(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "910c3c01-4ca5-445e-a7b6-3bb077997ecf",
   "metadata": {},
   "outputs": [],
   "source": [
    "ann_data.write_h5ad('/Users/hk9/tdt_datasets/Macaque_HMBA_basalganglia_AIT_pre-print_2.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1631544b-122a-412d-84c3-ac4c812535b0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "venv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
